# เอกสารประกอบการสอน Week 07 — Exception (ระดับพื้นฐาน)

เอกสารฉบับนี้ครอบคลุมหัวข้อที่ 1 ถึงหัวข้อที่ 5 ของสัปดาห์ที่ 7 เรื่องการจัดการข้อผิดพลาดขณะรัน (exception)
ในภาษาไพทอน เนื้อหาทั้งหมดในเอกสารฉบับนี้**ตรงกับขอบเขตการสอบจริง**ตามที่ปรากฏใน `slides/07_exception.md`
(สไลด์ 1–8) กล่าวคือ หัวข้อที่ 1–5 ในเอกสารนี้คือเนื้อหาเดียวกันกับที่ถูกขยายความให้ละเอียดขึ้น พร้อมตัวอย่าง
โค้ดที่รันจริงและยืนยันผลลัพธ์แล้วทุกตัวอย่าง เพื่อใช้ประกอบการเตรียมสอบปากเปล่า (Oral Defense) ในสัปดาห์สุดท้าย
ของวิชา Programming 1

แนวคิดหลักของแต่ละหัวข้อจะอธิบายในลักษณะที่ไม่ผูกติดกับภาษาโปรแกรมใดภาษาหนึ่ง (language-agnostic) ก่อน
แล้วจึงลงรายละเอียดวากยสัมพันธ์ (syntax) เฉพาะของภาษาไพทอนตามหลัง ตัวอย่างโค้ดทุกตัวอย่างมีองค์ประกอบครบ
5 ส่วนตามมาตรฐานของเอกสารชุดนี้ ได้แก่ (1) โค้ดพร้อมเลขบรรทัดกำกับ (2) ผลลัพธ์ที่รันจริงและยืนยันแล้ว
(3) ตารางการเปลี่ยนแปลงค่า (4) คำอธิบายภาพรวม และ (5) จุดสังเกต

ตัวอย่างโค้ดทั้งหมดในเอกสารนี้ถูกรันจริงด้วย Python 3.11.15 ก่อนนำผลลัพธ์มาบันทึกไว้ ไม่มีผลลัพธ์ใดที่มาจาก
การคาดเดาหรือคัดลอกจากแหล่งอ้างอิงโดยไม่ได้รันซ้ำ ในการแสดง traceback ของแต่ละตัวอย่าง เอกสารนี้แสดงชื่อไฟล์
ในบรรทัด `File "..."` แบบย่อ (เช่น `exercise.py`) แทนที่จะแสดงพาธเต็มของโฟลเดอร์ชั่วคราวที่ใช้ในการทดสอบ
ทั้งนี้เพื่อความอ่านง่าย โดยชนิดข้อผิดพลาด ข้อความอธิบาย หมายเลขบรรทัด และเครื่องหมายชี้ตำแหน่ง (`^`, `~`)
ทั้งหมดคงไว้ตรงตามผลลัพธ์จริงที่ได้จากการรันทุกประการ

---

## หัวข้อที่ 1: Exception คืออะไร ต่างจาก Syntax Error อย่างไร

### แนวคิด

ในการเขียนโปรแกรมด้วยภาษาใดก็ตามที่มีขั้นตอนการแปลหรือตีความคำสั่งก่อนการทำงานจริง ข้อผิดพลาดที่อาจเกิดขึ้น
สามารถแบ่งออกได้เป็นสองลักษณะใหญ่ตามช่วงเวลาที่ตรวจพบ ลักษณะแรกคือข้อผิดพลาดที่ตรวจพบได้จากการตรวจสอบ
โครงสร้างของคำสั่งเพียงอย่างเดียว โดยไม่จำเป็นต้องรันโปรแกรมเลยแม้แต่บรรทัดเดียว เรียกว่าข้อผิดพลาดทางไวยากรณ์
(syntax error) ลักษณะที่สองคือข้อผิดพลาดที่เกิดขึ้นระหว่างที่โปรแกรมกำลังทำงานอยู่จริง อันเป็นผลจากสถานการณ์
ที่ผู้เขียนโปรแกรมไม่ได้เตรียมรองรับไว้ล่วงหน้า เช่น การรับข้อมูลนำเข้าที่ไม่ถูกต้อง การเข้าถึงตำแหน่งข้อมูลที่ไม่มี
อยู่จริง หรือการดำเนินการกับชนิดข้อมูลที่ไม่สอดคล้องกัน ข้อผิดพลาดลักษณะนี้เรียกว่าข้อผิดพลาดขณะรัน (exception)
ซึ่งเป็นแนวคิดที่ปรากฏอยู่ในภาษาโปรแกรมสมัยใหม่แทบทุกภาษา ไม่ได้จำกัดเฉพาะภาษาใดภาษาหนึ่ง

ในภาษาไพทอน ข้อผิดพลาดทางไวยากรณ์จะถูกตรวจพบในขั้นตอนการแปลงโค้ดต้นฉบับ (parsing) ก่อนที่ตัวแปลภาษาจะเริ่ม
ประมวลผลคำสั่งใด ๆ เลย หากพบข้อผิดพลาดประเภทนี้ โปรแกรมจะไม่ทำงานแม้แต่บรรทัดเดียว ตัวอย่างของข้อผิดพลาด
ทางไวยากรณ์ที่พบบ่อยคือการเยื้องบรรทัด (indentation) ที่ไม่ถูกต้อง ซึ่งไพทอนจะรายงานเป็น `IndentationError`
อันเป็นชนิดย่อยของ `SyntaxError` ในทางตรงข้าม ข้อผิดพลาดขณะรันจะเกิดขึ้นหลังจากโปรแกรมเริ่มทำงานไปแล้วระยะหนึ่ง
และเมื่อเกิดขึ้น ไพทอนจะแสดงข้อความที่เรียกว่า traceback (ข้อความแสดงลำดับข้อผิดพลาด) ซึ่งระบุลำดับการเรียกใช้งาน
ของโปรแกรมจนถึงบรรทัดที่ทำให้เกิดข้อผิดพลาด พร้อมทั้งชนิดของข้อผิดพลาดและข้อความอธิบายเพิ่มเติม หากไม่มีการ
ดักจับข้อผิดพลาดนี้ไว้ โปรแกรมจะหยุดทำงานทันทีที่บรรทัดนั้น และคำสั่งที่อยู่ถัดจากจุดเกิดข้อผิดพลาดจะไม่ถูกประมวลผลเลย

### ตัวอย่าง Python

#### ตัวอย่างที่ 1.1 — ระบบคำนวณค่าเฉลี่ยคะแนนสอบ (แสดง exception ที่ไม่ถูกดักจับ)

In [1]:
print("เริ่มโปรแกรมคำนวณค่าเฉลี่ยคะแนนสอบ")
scores = [80, 90, 70]
total = sum(scores)
average = total / len(scores)
print("ค่าเฉลี่ยคือ", average)
print(scores[5])
print("บรรทัดนี้จะไม่ถูกรัน")

เริ่มโปรแกรมคำนวณค่าเฉลี่ยคะแนนสอบ
ค่าเฉลี่ยคือ 80.0


IndexError: list index out of range

**Output (รันจริงยืนยันแล้ว):**
```
เริ่มโปรแกรมคำนวณค่าเฉลี่ยคะแนนสอบ
ค่าเฉลี่ยคือ 80.0
Traceback (most recent call last):
  File "exercise.py", line 6, in <module>
    print(scores[5])
          ~~~~~~^^^
IndexError: list index out of range
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 01 | `print(...)` | - | พิมพ์ข้อความเริ่มต้นโปรแกรม |
| 02 | `scores = [80, 90, 70]` | `scores = [80, 90, 70]` | รายการมีสมาชิก 3 ตัว ดัชนีที่ใช้ได้คือ 0, 1, 2 เท่านั้น |
| 03 | `total = sum(scores)` | `total = 240` | รวมค่าทั้งหมดในรายการ |
| 04 | `average = total / len(scores)` | `average = 80.0` | หารด้วยจำนวนสมาชิกในรายการ (3) |
| 05 | `print(...)` | - | แสดงค่าเฉลี่ยที่คำนวณได้สำเร็จ |
| 06 | `print(scores[5])` | - | เข้าถึงดัชนี 5 ซึ่งไม่มีอยู่จริงในรายการ เกิด `IndexError` โปรแกรมหยุดทำงานทันที ณ บรรทัดนี้ |
| 07 | (ไม่ถูกรัน) | - | คำสั่งอยู่หลังจุดเกิดข้อผิดพลาด จึงไม่ถูกประมวลผลเลย |

**คำอธิบายภาพรวม:**

โปรแกรมนี้เริ่มต้นด้วยการคำนวณค่าเฉลี่ยของคะแนนสอบในบรรทัดที่ 2–4 ซึ่งทำงานได้สำเร็จโดยไม่มีปัญหาใด ๆ และ
พิมพ์ผลลัพธ์ออกมาในบรรทัดที่ 5 ได้ตามปกติ อย่างไรก็ตาม ในบรรทัดที่ 6 โปรแกรมพยายามเข้าถึงสมาชิกลำดับที่ 5
ของรายการ `scores` ซึ่งมีสมาชิกอยู่เพียง 3 ตัว (ดัชนี 0 ถึง 2) การกระทำเช่นนี้ทำให้ไพทอนสร้างข้อผิดพลาดขณะรัน
ชนิด `IndexError` ขึ้นมาทันที เนื่องจากโค้ดนี้ไม่มีกลไกใดดักจับข้อผิดพลาดไว้เลย ตัวแปลภาษาจึงพิมพ์ traceback
ออกมาแสดงตำแหน่งที่เกิดปัญหา แล้วยุติการทำงานของโปรแกรมทันที ส่งผลให้บรรทัดที่ 7 ไม่มีโอกาสถูกประมวลผลเลย

**จุดสังเกต:**

- บรรทัดที่ 5 (การพิมพ์ค่าเฉลี่ย) ยังคงทำงานได้ตามปกติ เพราะข้อผิดพลาดยังไม่เกิดขึ้น ณ จุดนั้นของโปรแกรม
- บรรทัดที่ 7 ซึ่งอยู่หลังจุดเกิด `IndexError` ไม่ถูกรันเลยแม้จะเป็นเพียงคำสั่ง `print` ธรรมดาที่ไม่เกี่ยวข้องกับ
  ตัวแปรที่มีปัญหา แสดงให้เห็นว่าเมื่อ exception เกิดขึ้นและไม่ถูกดักจับ การทำงานทั้งโปรแกรมจะหยุดทันที
  ไม่ใช่แค่หยุดเฉพาะคำสั่งที่มีปัญหา
- รายการ `scores` มีสมาชิก 3 ตัว การเรียก `scores[5]` จึงเกินขอบเขตดัชนีที่มีอยู่จริง ข้อผิดพลาดนี้เกิดขึ้นเฉพาะ
  เมื่อโปรแกรมทำงานมาถึงบรรทัดนั้นจริง ๆ ไม่สามารถตรวจพบได้ล่วงหน้าเพียงจากการอ่านโครงสร้างของโค้ด

#### ตัวอย่างที่ 1.2 — ระบบคำนวณราคารวมของสินค้าในตะกร้า (บริบทอื่นที่ต่างโดเมน)

In [2]:
print("เริ่มโปรแกรมคำนวณราคารวมของสินค้าในตะกร้า")
price = 150
shipping_fee = "50"
total_price = price + shipping_fee
print("ราคารวมคือ", total_price)

เริ่มโปรแกรมคำนวณราคารวมของสินค้าในตะกร้า


TypeError: unsupported operand type(s) for +: 'int' and 'str'

**Output (รันจริงยืนยันแล้ว):**
```
เริ่มโปรแกรมคำนวณราคารวมของสินค้าในตะกร้า
Traceback (most recent call last):
  File "exercise.py", line 4, in <module>
    total_price = price + shipping_fee
                  ~~~~~~^~~~~~~~~~~~~~
TypeError: unsupported operand type(s) for +: 'int' and 'str'
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 01 | `print(...)` | - | พิมพ์ข้อความเริ่มต้นโปรแกรม |
| 02 | `price = 150` | `price = 150` (ชนิด `int`) | ราคาสินค้าเป็นตัวเลขจำนวนเต็ม |
| 03 | `shipping_fee = "50"` | `shipping_fee = "50"` (ชนิด `str`) | ค่าจัดส่งถูกเก็บเป็นข้อความ ไม่ใช่ตัวเลข |
| 04 | `total_price = price + shipping_fee` | - | นำ `int` มาบวกกับ `str` โดยตรง เกิด `TypeError` โปรแกรมหยุดทำงานทันที |
| 05 | (ไม่ถูกรัน) | - | คำสั่งอยู่หลังจุดเกิดข้อผิดพลาด จึงไม่ถูกประมวลผลเลย |

**คำอธิบายภาพรวม:**

โปรแกรมนี้กำหนดราคาสินค้าในบรรทัดที่ 2 เป็นชนิดตัวเลขจำนวนเต็ม (`int`) และกำหนดค่าจัดส่งในบรรทัดที่ 3 เป็น
ชนิดข้อความ (`str`) โดยไม่ได้ตั้งใจ เมื่อโปรแกรมพยายามนำค่าทั้งสองมาบวกกันในบรรทัดที่ 4 ไพทอนไม่สามารถ
ดำเนินการบวกระหว่างชนิดข้อมูล `int` กับ `str` ได้โดยอัตโนมัติ จึงเกิดข้อผิดพลาดขณะรันชนิด `TypeError` ขึ้น
เนื่องจากไม่มีการดักจับข้อผิดพลาดไว้ โปรแกรมจึงหยุดทำงานทันทีที่บรรทัดนี้ และบรรทัดที่ 5 ซึ่งควรจะพิมพ์ราคารวม
จึงไม่มีโอกาสถูกประมวลผลเลย

**จุดสังเกต:**

- `TypeError` ในตัวอย่างนี้เกิดจากการนำชนิดข้อมูล `int` และ `str` มาบวกกันโดยตรง ซึ่งไพทอนไม่แปลงชนิดข้อมูล
  ให้อัตโนมัติในกรณีของตัวดำเนินการ `+` ระหว่างตัวเลขกับข้อความ
- ข้อความ traceback ระบุชนิดข้อมูลทั้งสองฝั่งที่ขัดแย้งกันไว้อย่างชัดเจน (`'int'` และ `'str'`) ซึ่งเป็นข้อมูลที่มีประโยชน์
  มากในการวินิจฉัยว่าปัญหาคืออะไร
- ต่างจากตัวอย่างที่ 1.1 ตรงที่ต้นเหตุของข้อผิดพลาดในที่นี้คือชนิดข้อมูลไม่สอดคล้องกัน ไม่ใช่การเข้าถึงตำแหน่งดัชนี
  ที่เกินขอบเขต แต่ทั้งสองกรณีมีลักษณะร่วมกันคือเป็นข้อผิดพลาดขณะรันที่ตรวจพบได้เฉพาะเมื่อโปรแกรมทำงานถึง
  บรรทัดนั้นจริง

#### ตัวอย่างที่ 1.3 — ระบบตรวจสอบอายุผู้สมัครสมาชิก (ข้อผิดพลาดทางไวยากรณ์)

In [3]:
age = 20
    print("อายุคือ", age)

IndentationError: unexpected indent (3066929110.py, line 2)

**Output (รันจริงยืนยันแล้ว):**
```
  File "exercise.py", line 2
    print("อายุคือ", age)
IndentationError: unexpected indent
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 01 | (ไม่ถูกรัน) | - | ไม่มีบรรทัดใดถูกรันเลยแม้แต่บรรทัดเดียว เนื่องจากข้อผิดพลาดถูกตรวจพบตั้งแต่ขั้นตอนตรวจสอบไวยากรณ์ ก่อนที่โปรแกรมจะเริ่มทำงานจริง |
| 02 | (ไม่ถูกรัน) | - | บรรทัดนี้มีการเยื้องบรรทัด (indentation) ที่ไม่สอดคล้องกับโครงสร้างของโปรแกรม ทำให้ตัวแปลภาษาปฏิเสธโค้ดทั้งไฟล์ตั้งแต่ขั้นตอนตรวจสอบไวยากรณ์ |

**คำอธิบายภาพรวม:**

โค้ดในตัวอย่างนี้มีเพียง 2 บรรทัด บรรทัดที่ 1 กำหนดค่าตัวแปร `age` ตามปกติ แต่บรรทัดที่ 2 มีการเยื้องบรรทัดด้วย
ช่องว่างนำหน้าโดยไม่มีเหตุผลทางโครงสร้างใด ๆ รองรับ (ไม่ได้อยู่ภายใต้คำสั่งที่ต้องมีการเยื้อง เช่น `if`, `for`, หรือ
`def`) เมื่อไพทอนตรวจสอบโครงสร้างของไฟล์ทั้งหมดก่อนเริ่มทำงาน จะพบว่าการเยื้องบรรทัดนี้ไม่ถูกต้องตามหลัก
ไวยากรณ์ของภาษา จึงรายงานเป็น `IndentationError` ซึ่งเป็นชนิดย่อยของ `SyntaxError` และปฏิเสธที่จะรันโค้ดทั้งไฟล์
เลยแม้แต่บรรทัดเดียว ผลลัพธ์ที่ได้จึงไม่มีข้อความใด ๆ จากคำสั่ง `print` ปรากฏออกมาเลย มีเพียงข้อความรายงาน
ข้อผิดพลาดทางไวยากรณ์เท่านั้น

**จุดสังเกต:**

- ไม่มีข้อความใด ๆ จากคำสั่ง `print` ปรากฏบนหน้าจอเลยแม้แต่บรรทัดเดียว เพราะข้อผิดพลาดถูกตรวจพบตั้งแต่ขั้นตอน
  ตรวจสอบไวยากรณ์ ก่อนที่โปรแกรมจะเริ่มทำงานจริง
- ข้อความรายงานคือ `IndentationError` ซึ่งเป็นชนิดย่อยของ `SyntaxError` ไม่ใช่ข้อผิดพลาดขณะรัน (exception)
  จึงไม่สามารถใช้ `try`/`except` ดักจับข้อผิดพลาดชนิดนี้ได้ เพราะโปรแกรมยังไม่ทันเริ่มทำงานเลยด้วยซ้ำ
- สังเกตความแตกต่างจากตัวอย่างที่ 1.1 และ 1.2 ที่โปรแกรมยังพิมพ์ข้อความบางส่วนออกมาได้ก่อนจะพบข้อผิดพลาด
  (เพราะเป็น exception ที่เกิดระหว่างการรัน) ในขณะที่ตัวอย่างนี้ไม่มีข้อความใดถูกพิมพ์ออกมาเลย

### ขอบเขตเนื้อหา

หัวข้อนี้ตรงกับเนื้อหาที่สอนใน Slide 2 ของ `slides/07_exception.md`

### เชื่อมโยงกับหัวข้ออื่น

หัวข้อถัดไป (หัวข้อที่ 2) จะแสดงวิธีใช้ `try`/`except` เพื่อดักจับข้อผิดพลาดขณะรันที่กล่าวถึงในตัวอย่างที่ 1.1
และ 1.2 ไม่ให้โปรแกรมหยุดทำงานกลางคัน ส่วนวิธีจัดการกับข้อผิดพลาดทางไวยากรณ์อย่างตัวอย่างที่ 1.3 นั้นไม่สามารถ
ทำได้ด้วย `try`/`except` เนื่องจากต้องแก้ไขที่ตัวโค้ดต้นฉบับโดยตรงก่อนโปรแกรมจะเริ่มทำงาน

---

## หัวข้อที่ 2: try/except พื้นฐาน (จับ exception ประเภทเดียว)

### แนวคิด

แนวคิดพื้นฐานของการจัดการข้อผิดพลาดขณะรันในภาษาโปรแกรมส่วนใหญ่คือการแบ่งโค้ดออกเป็นสองส่วน ส่วนแรกคือ
ส่วนที่ "ทดลองรัน" คำสั่งที่อาจเกิดปัญหาได้ และส่วนที่สองคือส่วนที่ "รับมือ" เมื่อปัญหานั้นเกิดขึ้นจริง หากคำสั่งใน
ส่วนแรกทำงานได้สำเร็จโดยไม่มีปัญหา ส่วนที่สองก็จะไม่ถูกเรียกใช้เลย แต่หากเกิดปัญหาขึ้น ณ บรรทัดใดบรรทัดหนึ่ง
ในส่วนแรก การทำงานจะกระโดดออกจากส่วนแรกทันที ไปยังส่วนที่สองที่เตรียมไว้สำหรับรับมือกับปัญหานั้นโดยเฉพาะ
คำสั่งที่เหลืออยู่ในส่วนแรกหลังจากจุดเกิดปัญหาจะไม่ถูกประมวลผลอีกเลย

ในภาษาไพทอน กลไกนี้เขียนด้วยคำสั่ง `try` (บล็อกทดลองรัน) ตามด้วยคำสั่ง `except` (บล็อกดักข้อผิดพลาด) โดย
ระบุชื่อชนิดของข้อผิดพลาดที่ต้องการดักจับต่อท้าย `except` เมื่อคำสั่งภายในบล็อก `try` ทำงานแล้วเกิดข้อผิดพลาด
ที่มีชนิดตรงกับที่ระบุไว้ใน `except` การทำงานจะกระโดดเข้าไปในบล็อก `except` ทันที และคำสั่งที่เหลือในบล็อก
`try` จะไม่ถูกรันเลย เมื่อบล็อก `except` ทำงานเสร็จแล้ว โปรแกรมจะดำเนินต่อไปยังคำสั่งถัดจากทั้งโครงสร้าง
`try`/`except` ตามปกติ ไม่หยุดทำงานทั้งโปรแกรมเหมือนกรณีที่ไม่มีการดักจับ

### ตัวอย่าง Python

#### ตัวอย่างที่ 2.1 — ระบบคำนวณปริมาณน้ำมันเฉลี่ยต่อถัง

In [4]:
total_liters = 0
number_of_tanks = 0
try:
    average_per_tank = total_liters / number_of_tanks
    print("น้ำมันเฉลี่ยต่อถังคือ", average_per_tank, "ลิตร")
except ZeroDivisionError:
    print("ไม่สามารถคำนวณค่าเฉลี่ยได้ เนื่องจากจำนวนถังเป็นศูนย์")
print("โปรแกรมทำงานต่อไปได้ตามปกติ")

ไม่สามารถคำนวณค่าเฉลี่ยได้ เนื่องจากจำนวนถังเป็นศูนย์
โปรแกรมทำงานต่อไปได้ตามปกติ


**Output (รันจริงยืนยันแล้ว):**
```
ไม่สามารถคำนวณค่าเฉลี่ยได้ เนื่องจากจำนวนถังเป็นศูนย์
โปรแกรมทำงานต่อไปได้ตามปกติ
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 01 | `total_liters = 0` | `total_liters = 0` | กำหนดปริมาณน้ำมันรวมเป็นศูนย์ |
| 02 | `number_of_tanks = 0` | `number_of_tanks = 0` | กำหนดจำนวนถังเป็นศูนย์ |
| 03 | `try:` | - | เริ่มบล็อกทดลองรัน |
| 04 | `average_per_tank = total_liters / number_of_tanks` | - | คำนวณ `0 / 0` เกิด `ZeroDivisionError` ทันที กระโดดออกจากบล็อก `try` |
| 05 | (ไม่ถูกรัน) | - | อยู่หลังจุดเกิดข้อผิดพลาดในบล็อก `try` จึงไม่ถูกประมวลผล |
| 06 | `except ZeroDivisionError:` | - | ชนิดข้อผิดพลาดตรงกับที่ระบุไว้ การทำงานกระโดดเข้าบล็อกนี้ |
| 07 | `print(...)` | - | พิมพ์ข้อความแจ้งเตือนที่เตรียมไว้สำหรับกรณีนี้ |
| 08 | `print(...)` | - | โปรแกรมทำงานต่อได้ตามปกติหลังจบโครงสร้าง `try`/`except` |

**คำอธิบายภาพรวม:**

โปรแกรมนี้จำลองสถานการณ์ที่ปริมาณน้ำมันรวม (`total_liters`) และจำนวนถัง (`number_of_tanks`) ถูกกำหนดเป็น
ศูนย์ทั้งคู่ในบรรทัดที่ 1–2 เมื่อโปรแกรมพยายามคำนวณค่าเฉลี่ยต่อถังในบรรทัดที่ 4 ซึ่งอยู่ภายในบล็อก `try` จะเกิด
การหารด้วยศูนย์ ทำให้เกิดข้อผิดพลาดขณะรันชนิด `ZeroDivisionError` ทันที เนื่องจากโครงสร้าง `try`/`except`
ในบรรทัดที่ 3–7 ได้เตรียมดักจับข้อผิดพลาดชนิดนี้ไว้แล้ว การทำงานจึงกระโดดออกจากบล็อก `try` (บรรทัดที่ 5
จึงไม่ถูกรัน) ไปยังบล็อก `except` ในบรรทัดที่ 6–7 ทันที ซึ่งพิมพ์ข้อความแจ้งเตือนที่เตรียมไว้ เมื่อบล็อก `except`
ทำงานเสร็จ โปรแกรมจึงดำเนินต่อไปยังบรรทัดที่ 8 ได้ตามปกติ โดยไม่หยุดทำงานทั้งโปรแกรมแต่อย่างใด

**จุดสังเกต:**

- บรรทัดที่ 8 ยังคงทำงานได้ตามปกติ ต่างจากตัวอย่างที่ 1.1 และ 1.2 ที่โปรแกรมหยุดทำงานกลางคันเพราะไม่มีการ
  ดักจับข้อผิดพลาดไว้เลย นี่คือประโยชน์หลักของ `try`/`except`
- บรรทัดที่ 5 ซึ่งอยู่หลังจุดเกิดข้อผิดพลาดภายในบล็อก `try` เดียวกัน ก็ไม่ถูกรันเช่นกัน แม้จะมีบล็อก `except`
  รองรับอยู่แล้วก็ตาม แสดงว่าการดักจับข้อผิดพลาดไม่ได้ทำให้บรรทัดที่เหลือใน `try` กลับมาทำงานต่อได้
- ชื่อชนิดข้อผิดพลาดที่ระบุไว้หลัง `except` (`ZeroDivisionError`) ต้องตรงกับชนิดที่เกิดขึ้นจริงจึงจะดักจับได้ หากระบุ
  ชนิดอื่นที่ไม่ตรงกัน โปรแกรมจะไม่ถูกดักจับ ดังจะเห็นตัวอย่างเพิ่มเติมได้ในหัวข้อที่ 4

#### ตัวอย่างที่ 2.2 — ระบบยืมหนังสือของห้องสมุด (บริบทอื่นที่ต่างโดเมน)

In [5]:
book_quantity_text = "ห้าเล่ม"
try:
    book_quantity = int(book_quantity_text)
    print("จำนวนหนังสือที่ยืมคือ", book_quantity, "เล่ม")
except ValueError:
    print("ข้อมูลจำนวนหนังสือไม่ถูกต้อง กรุณากรอกเป็นตัวเลข")
print("ระบบยืมหนังสือพร้อมรับคำสั่งถัดไป")

ข้อมูลจำนวนหนังสือไม่ถูกต้อง กรุณากรอกเป็นตัวเลข
ระบบยืมหนังสือพร้อมรับคำสั่งถัดไป


**Output (รันจริงยืนยันแล้ว):**
```
ข้อมูลจำนวนหนังสือไม่ถูกต้อง กรุณากรอกเป็นตัวเลข
ระบบยืมหนังสือพร้อมรับคำสั่งถัดไป
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 01 | `book_quantity_text = "ห้าเล่ม"` | `book_quantity_text = "ห้าเล่ม"` | ข้อความนี้ไม่ใช่ตัวเลขล้วน จึงแปลงเป็นจำนวนเต็มไม่ได้ |
| 02 | `try:` | - | เริ่มบล็อกทดลองรัน |
| 03 | `book_quantity = int(book_quantity_text)` | - | พยายามแปลงข้อความเป็นจำนวนเต็ม เกิด `ValueError` ทันที |
| 04 | (ไม่ถูกรัน) | - | อยู่หลังจุดเกิดข้อผิดพลาดในบล็อก `try` จึงไม่ถูกประมวลผล |
| 05 | `except ValueError:` | - | ชนิดข้อผิดพลาดตรงกับที่ระบุไว้ การทำงานกระโดดเข้าบล็อกนี้ |
| 06 | `print(...)` | - | พิมพ์ข้อความแจ้งเตือนที่เตรียมไว้สำหรับกรณีนี้ |
| 07 | `print(...)` | - | โปรแกรมทำงานต่อได้ตามปกติหลังจบโครงสร้าง `try`/`except` |

**คำอธิบายภาพรวม:**

ตัวอย่างนี้จำลองระบบยืมหนังสือของห้องสมุดที่รับจำนวนหนังสือมาเป็นข้อความ (`book_quantity_text`) ในบรรทัดที่ 1
ซึ่งเป็นคำว่า "ห้าเล่ม" อันเป็นข้อความภาษาไทย ไม่ใช่ตัวเลขล้วน เมื่อโปรแกรมพยายามแปลงข้อความนี้ให้เป็นจำนวนเต็ม
ด้วยฟังก์ชัน `int(...)` ในบรรทัดที่ 3 ซึ่งอยู่ภายในบล็อก `try` จะเกิดข้อผิดพลาดขณะรันชนิด `ValueError` ทันที
เนื่องจากบรรทัดที่ 5–6 ได้เตรียมดักจับข้อผิดพลาดชนิดนี้ไว้แล้ว การทำงานจึงกระโดดเข้าบล็อก `except` ทันที
(บรรทัดที่ 4 จึงไม่ถูกรัน) และพิมพ์ข้อความแจ้งเตือนว่าข้อมูลไม่ถูกต้อง จากนั้นโปรแกรมจึงดำเนินต่อไปยังบรรทัดที่ 7
ได้ตามปกติ

**จุดสังเกต:**

- ต่างจากตัวอย่างที่ 2.1 ตรงที่ชนิดของข้อผิดพลาดในที่นี้คือ `ValueError` (เกิดจากการแปลงชนิดข้อมูลไม่สำเร็จ)
  ไม่ใช่ `ZeroDivisionError` แสดงว่าโครงสร้าง `try`/`except` เดียวกันสามารถนำไปใช้ดักจับข้อผิดพลาดคนละชนิด
  ได้ เพียงแค่เปลี่ยนชื่อชนิดที่ระบุไว้หลัง `except`
- หากผู้ใช้กรอกข้อมูลเป็นตัวเลขล้วน เช่น `"5"` แทนที่จะเป็น `"ห้าเล่ม"` บล็อก `try` จะทำงานสำเร็จทั้งหมด และ
  บล็อก `except` จะไม่ถูกเรียกใช้เลย
- ฟังก์ชัน `int(...)` เป็นจุดที่มักเกิด `ValueError` บ่อยที่สุดในโปรแกรมที่รับข้อมูลนำเข้าจากผู้ใช้ เพราะไม่สามารถ
  ควบคุมได้ว่าผู้ใช้จะพิมพ์อะไรเข้ามา

#### ตัวอย่างที่ 2.3 — ระบบพยากรณ์อากาศรายวัน (edge case: ดัชนีเกินขอบเขต)

In [6]:
weekly_temperatures = [31, 32, 30, 29]
try:
    friday_temperature = weekly_temperatures[4]
    print("อุณหภูมิวันศุกร์คือ", friday_temperature, "องศาเซลเซียส")
except IndexError:
    print("ไม่มีข้อมูลอุณหภูมิของวันที่ระบุ")
print("ระบบพยากรณ์อากาศทำงานต่อไปได้ตามปกติ")

ไม่มีข้อมูลอุณหภูมิของวันที่ระบุ
ระบบพยากรณ์อากาศทำงานต่อไปได้ตามปกติ


**Output (รันจริงยืนยันแล้ว):**
```
ไม่มีข้อมูลอุณหภูมิของวันที่ระบุ
ระบบพยากรณ์อากาศทำงานต่อไปได้ตามปกติ
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 01 | `weekly_temperatures = [31, 32, 30, 29]` | `weekly_temperatures` มีสมาชิก 4 ตัว | ดัชนีที่ใช้ได้คือ 0, 1, 2, 3 เท่านั้น |
| 02 | `try:` | - | เริ่มบล็อกทดลองรัน |
| 03 | `friday_temperature = weekly_temperatures[4]` | - | เข้าถึงดัชนี 4 ซึ่งไม่มีอยู่จริง เกิด `IndexError` ทันที |
| 04 | (ไม่ถูกรัน) | - | อยู่หลังจุดเกิดข้อผิดพลาดในบล็อก `try` จึงไม่ถูกประมวลผล |
| 05 | `except IndexError:` | - | ชนิดข้อผิดพลาดตรงกับที่ระบุไว้ การทำงานกระโดดเข้าบล็อกนี้ |
| 06 | `print(...)` | - | พิมพ์ข้อความแจ้งเตือนที่เตรียมไว้สำหรับกรณีนี้ |
| 07 | `print(...)` | - | โปรแกรมทำงานต่อได้ตามปกติหลังจบโครงสร้าง `try`/`except` |

**คำอธิบายภาพรวม:**

ตัวอย่างนี้จำลองระบบพยากรณ์อากาศที่เก็บอุณหภูมิรายวันของสัปดาห์ไว้ในรายการ `weekly_temperatures` ซึ่งมี
สมาชิกเพียง 4 ตัว (แทนวันจันทร์ถึงพฤหัสบดี) ในบรรทัดที่ 1 เมื่อโปรแกรมพยายามเข้าถึงอุณหภูมิของ "วันศุกร์" ที่
ตำแหน่งดัชนี 4 ในบรรทัดที่ 3 ซึ่งอยู่ภายในบล็อก `try` จะพบว่าตำแหน่งนี้ไม่มีอยู่จริงในรายการ จึงเกิดข้อผิดพลาด
ขณะรันชนิด `IndexError` ทันที เนื่องจากบรรทัดที่ 5–6 ได้เตรียมดักจับข้อผิดพลาดชนิดนี้ไว้แล้ว การทำงานจึงกระโดด
เข้าบล็อก `except` ทันที (บรรทัดที่ 4 จึงไม่ถูกรัน) และพิมพ์ข้อความแจ้งเตือนว่าไม่มีข้อมูล จากนั้นโปรแกรมจึง
ดำเนินต่อไปยังบรรทัดที่ 7 ได้ตามปกติ

**จุดสังเกต:**

- ข้อผิดพลาดชนิด `IndexError` ในที่นี้เกิดจากความไม่สอดคล้องระหว่างจำนวนสมาชิกจริงในรายการ (4 ตัว) กับ
  ตำแหน่งดัชนีที่พยายามเข้าถึง (ดัชนี 4 ซึ่งเป็นตำแหน่งที่ 5) เป็นความผิดพลาดที่พบบ่อยเมื่อผู้เขียนโปรแกรมนับ
  จำนวนสมาชิกผิดหรือลืมว่าดัชนีเริ่มต้นที่ 0 ไม่ใช่ 1
- ทั้งสามตัวอย่างในหัวข้อนี้ (2.1, 2.2, 2.3) มีโครงสร้างเดียวกันทุกประการ คือ `try` หนึ่งบล็อกคู่กับ `except`
  หนึ่งบล็อกที่ระบุชนิดข้อผิดพลาดเพียงชนิดเดียว ต่างกันเพียงชนิดข้อผิดพลาดและโดเมนของปัญหาเท่านั้น แสดงว่า
  รูปแบบ `try`/`except` พื้นฐานนี้นำไปประยุกต์ใช้ได้กับสถานการณ์ที่หลากหลาย
- หากเปลี่ยนดัชนีในบรรทัดที่ 3 จาก `4` เป็น `3` (ตำแหน่งสุดท้ายที่มีอยู่จริง) บล็อก `try` จะทำงานสำเร็จทั้งหมด
  และบล็อก `except` จะไม่ถูกเรียกใช้เลย

### ขอบเขตเนื้อหา

หัวข้อนี้ตรงกับเนื้อหาที่สอนใน Slide 3 ของ `slides/07_exception.md`

### เชื่อมโยงกับหัวข้ออื่น

โครงสร้าง `try`/`except` ในหัวข้อนี้เป็นพื้นฐานของหัวข้อที่ 4 (การจับหลาย exception ในฟังก์ชันเดียว) และหัวข้อที่ 5
(การเขียนฟังก์ชันเชิงป้องกัน) ซึ่งทั้งสองหัวข้อนำรูปแบบเดียวกันนี้ไปประยุกต์ใช้ภายในฟังก์ชัน ส่วนชนิดข้อผิดพลาดทั้ง 3
ที่ปรากฏในตัวอย่างของหัวข้อนี้ (`ZeroDivisionError`, `ValueError`, `IndexError`) จะถูกอธิบายอย่างเป็นระบบร่วมกับ
`TypeError` ในหัวข้อที่ 3 ถัดไป

---

## หัวข้อที่ 3: Exception types ที่พบบ่อย (ZeroDivisionError, ValueError, TypeError, IndexError)

### แนวคิด

แนวคิดเรื่องข้อผิดพลาดขณะรันไม่ได้มีเพียงชนิดเดียว ในทุกภาษาโปรแกรมที่รองรับการจัดการข้อผิดพลาด มักจะมีการ
จำแนกข้อผิดพลาดออกเป็นหลายชนิดตามลักษณะของสาเหตุที่ทำให้เกิดปัญหา เพื่อให้ผู้เขียนโปรแกรมสามารถเขียนโค้ด
รับมือกับแต่ละสาเหตุได้อย่างเจาะจง แทนที่จะต้องรับมือกับข้อผิดพลาดทุกชนิดด้วยวิธีเดียวกันทั้งหมด การรู้จักชื่อและ
สาเหตุของข้อผิดพลาดแต่ละชนิดจึงเป็นพื้นฐานสำคัญก่อนที่จะเขียนโค้ดดักจับข้อผิดพลาดได้อย่างถูกต้อง

ในภาษาไพทอน มีชนิดข้อผิดพลาดที่กำหนดไว้ในตัวภาษาอยู่จำนวนมาก แต่ในระดับเนื้อหาของสัปดาห์นี้จะเน้นเพียง
4 ชนิดที่พบบ่อยที่สุดในการเขียนโปรแกรมเบื้องต้น ได้แก่

| ชนิดข้อผิดพลาด | เกิดขึ้นเมื่อใด |
|---|---|
| `ZeroDivisionError` | มีการหารด้วยศูนย์ (ทั้งการหารปกติ `/` และการหารเอาเศษ `%`) |
| `ValueError` | มีการแปลงชนิดข้อมูล (เช่น `int(...)`, `float(...)`) ด้วยค่าที่ไม่สามารถแปลงได้ตามชนิดที่ต้องการ แม้ชนิดข้อมูลตั้งต้นจะถูกต้อง |
| `TypeError` | มีการดำเนินการ (เช่น การบวก การคูณ) ระหว่างชนิดข้อมูลที่ไม่สอดคล้องกัน หรือเรียกใช้ฟังก์ชันด้วยชนิดข้อมูลที่ไม่รองรับ |
| `IndexError` | มีการเข้าถึงตำแหน่งดัชนีของรายการ (list) ที่ไม่มีอยู่จริง |

ทั้ง 4 ชนิดนี้เป็นเพียงตัวอย่างส่วนหนึ่งจากชนิดข้อผิดพลาดมาตรฐานที่มีอยู่ทั้งหมดในภาษาไพทอน แต่เป็นชนิดที่พบบ่อย
ที่สุดในโปรแกรมระดับเบื้องต้นที่เกี่ยวข้องกับการคำนวณตัวเลขและการจัดการรายการข้อมูล

### ตัวอย่าง Python

#### ตัวอย่างที่ 3.1 — ระบบตัดเกรดนักเรียน

In [7]:
print("สาธิตข้อผิดพลาดที่พบบ่อยในระบบตัดเกรดนักเรียน")

scores = []
try:
    average = sum(scores) / len(scores)
    print("ค่าเฉลี่ยคือ", average)
except ZeroDivisionError:
    print("ZeroDivisionError: ไม่มีคะแนนให้คำนวณค่าเฉลี่ย")

grade_text = "A+"
try:
    grade_number = int(grade_text)
    print("เกรดตัวเลขคือ", grade_number)
except ValueError:
    print("ValueError: ไม่สามารถแปลงตัวอักษรเกรดเป็นตัวเลขได้")

bonus_score = "5"
base_score = 70
try:
    final_score = base_score + bonus_score
    print("คะแนนรวมคือ", final_score)
except TypeError:
    print("TypeError: ไม่สามารถบวกตัวเลขกับข้อความได้โดยตรง")

student_grades = [85, 90, 78]
try:
    fourth_grade = student_grades[3]
    print("เกรดคนที่สี่คือ", fourth_grade)
except IndexError:
    print("IndexError: ไม่มีนักเรียนลำดับนี้ในรายชื่อ")

สาธิตข้อผิดพลาดที่พบบ่อยในระบบตัดเกรดนักเรียน
ZeroDivisionError: ไม่มีคะแนนให้คำนวณค่าเฉลี่ย
ValueError: ไม่สามารถแปลงตัวอักษรเกรดเป็นตัวเลขได้
TypeError: ไม่สามารถบวกตัวเลขกับข้อความได้โดยตรง
IndexError: ไม่มีนักเรียนลำดับนี้ในรายชื่อ


**Output (รันจริงยืนยันแล้ว):**
```
สาธิตข้อผิดพลาดที่พบบ่อยในระบบตัดเกรดนักเรียน
ZeroDivisionError: ไม่มีคะแนนให้คำนวณค่าเฉลี่ย
ValueError: ไม่สามารถแปลงตัวอักษรเกรดเป็นตัวเลขได้
TypeError: ไม่สามารถบวกตัวเลขกับข้อความได้โดยตรง
IndexError: ไม่มีนักเรียนลำดับนี้ในรายชื่อ
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 03 | `scores = []` | `scores = []` | รายการว่างเปล่า ไม่มีสมาชิก |
| 05 | `average = sum(scores) / len(scores)` | - | `sum([])` ได้ 0, `len([])` ได้ 0 คำนวณ `0 / 0` เกิด `ZeroDivisionError` |
| 08 | `print(...)` | - | ดักจับ `ZeroDivisionError` ได้สำเร็จ |
| 10 | `grade_text = "A+"` | `grade_text = "A+"` | ข้อความนี้ไม่ใช่ตัวเลข |
| 12 | `grade_number = int(grade_text)` | - | `int("A+")` แปลงไม่ได้ เกิด `ValueError` |
| 15 | `print(...)` | - | ดักจับ `ValueError` ได้สำเร็จ |
| 17–18 | กำหนดค่า `bonus_score`, `base_score` | `bonus_score = "5"` (str), `base_score = 70` (int) | ชนิดข้อมูลต่างกัน |
| 20 | `final_score = base_score + bonus_score` | - | บวก `int` กับ `str` โดยตรง เกิด `TypeError` |
| 23 | `print(...)` | - | ดักจับ `TypeError` ได้สำเร็จ |
| 25 | `student_grades = [85, 90, 78]` | สมาชิก 3 ตัว | ดัชนีที่ใช้ได้คือ 0, 1, 2 |
| 27 | `fourth_grade = student_grades[3]` | - | เข้าถึงดัชนี 3 ซึ่งไม่มีอยู่จริง เกิด `IndexError` |
| 30 | `print(...)` | - | ดักจับ `IndexError` ได้สำเร็จ |

**คำอธิบายภาพรวม:**

โปรแกรมนี้จัดโครงสร้างเป็น 4 ส่วนย่อยที่เป็นอิสระจากกัน แต่ละส่วนใช้โครงสร้าง `try`/`except` ของตัวเอง เพื่อสาธิต
ให้เห็นสาเหตุการเกิดข้อผิดพลาดแต่ละชนิดในบริบทเดียวกันคือระบบตัดเกรดนักเรียน ส่วนแรก (บรรทัดที่ 3–8) แสดงให้
เห็นว่าเมื่อรายการคะแนนว่างเปล่า การคำนวณค่าเฉลี่ยจะนำไปสู่การหารด้วยศูนย์ ทำให้เกิด `ZeroDivisionError`
ส่วนที่สอง (บรรทัดที่ 10–15) แสดงให้เห็นว่าการพยายามแปลงข้อความที่ไม่ใช่ตัวเลข (`"A+"`) ให้เป็นจำนวนเต็มด้วย
`int(...)` ทำให้เกิด `ValueError` ส่วนที่สาม (บรรทัดที่ 17–23) แสดงให้เห็นว่าการนำชนิดข้อมูล `int` และ `str`
มาบวกกันโดยตรงทำให้เกิด `TypeError` และส่วนสุดท้าย (บรรทัดที่ 25–30) แสดงให้เห็นว่าการเข้าถึงดัชนีที่เกิน
ขอบเขตของรายการทำให้เกิด `IndexError` แต่ละส่วนดักจับข้อผิดพลาดของตนเองได้สำเร็จ โปรแกรมจึงทำงานจนจบ
โดยไม่หยุดกลางคัน

**จุดสังเกต:**

- แม้ทั้ง 4 ส่วนจะอยู่ในบริบทเดียวกันคือระบบตัดเกรด แต่สาเหตุการเกิดข้อผิดพลาดแตกต่างกันโดยสิ้นเชิง การรู้จัก
  ชื่อและสาเหตุเฉพาะของแต่ละชนิดช่วยให้วินิจฉัยปัญหาได้เร็วขึ้นเมื่อเห็นชื่อข้อผิดพลาดใน traceback
- `ValueError` (บรรทัดที่ 10–15) กับ `TypeError` (บรรทัดที่ 17–23) มักถูกสับสนกัน เพราะทั้งคู่เกี่ยวข้องกับชนิด
  ข้อมูล แต่ต่างกันตรงที่ `ValueError` เกิดเมื่อชนิดข้อมูลถูกต้องแต่ค่าไม่เหมาะสม (เช่น ข้อความที่ไม่ใช่ตัวเลข)
  ในขณะที่ `TypeError` เกิดเมื่อนำชนิดข้อมูลที่ไม่รองรับการดำเนินการนั้นมาปะปนกันโดยตรง
- ทุกส่วนในตัวอย่างนี้พิมพ์ชื่อชนิดข้อผิดพลาดกำกับไว้ในข้อความ (เช่น `"ZeroDivisionError: ..."`) เพื่อให้เห็นชัดเจน
  ว่าข้อความแต่ละบรรทัดของผลลัพธ์สอดคล้องกับข้อผิดพลาดชนิดใด ซึ่งเป็นแนวทางการเขียนโค้ดสาธิตเพื่อการเรียน
  เท่านั้น ไม่ใช่รูปแบบการพิมพ์ข้อความที่จำเป็นต้องทำในโปรแกรมจริงเสมอไป

#### ตัวอย่างที่ 3.2 — ระบบสต็อกสินค้าร้านค้า (บริบทอื่นที่ต่างโดเมน)

In [8]:
print("สาธิตข้อผิดพลาดที่พบบ่อยในระบบสต็อกสินค้าร้านค้า")

sold_units = 40
number_of_branches = 0
try:
    average_sold = sold_units / number_of_branches
    print("ยอดขายเฉลี่ยต่อสาขาคือ", average_sold)
except ZeroDivisionError:
    print("ZeroDivisionError: ไม่มีจำนวนสาขาให้คำนวณค่าเฉลี่ย")

stock_input = "สามสิบ"
try:
    stock_count = int(stock_input)
    print("จำนวนสต็อกคือ", stock_count)
except ValueError:
    print("ValueError: ไม่สามารถแปลงข้อความเป็นจำนวนสต็อกได้")

unit_price = 25
discount_code = "SALE10"
try:
    discounted_price = unit_price + discount_code
    print("ราคาหลังหักส่วนลดคือ", discounted_price)
except TypeError:
    print("TypeError: ไม่สามารถบวกราคาด้วยรหัสส่วนลดที่เป็นข้อความได้")

warehouse_shelves = ["A1", "A2", "A3"]
try:
    fourth_shelf = warehouse_shelves[3]
    print("ชั้นวางที่สี่คือ", fourth_shelf)
except IndexError:
    print("IndexError: ไม่มีชั้นวางลำดับนี้ในคลังสินค้า")

สาธิตข้อผิดพลาดที่พบบ่อยในระบบสต็อกสินค้าร้านค้า
ZeroDivisionError: ไม่มีจำนวนสาขาให้คำนวณค่าเฉลี่ย
ValueError: ไม่สามารถแปลงข้อความเป็นจำนวนสต็อกได้
TypeError: ไม่สามารถบวกราคาด้วยรหัสส่วนลดที่เป็นข้อความได้
IndexError: ไม่มีชั้นวางลำดับนี้ในคลังสินค้า


**Output (รันจริงยืนยันแล้ว):**
```
สาธิตข้อผิดพลาดที่พบบ่อยในระบบสต็อกสินค้าร้านค้า
ZeroDivisionError: ไม่มีจำนวนสาขาให้คำนวณค่าเฉลี่ย
ValueError: ไม่สามารถแปลงข้อความเป็นจำนวนสต็อกได้
TypeError: ไม่สามารถบวกราคาด้วยรหัสส่วนลดที่เป็นข้อความได้
IndexError: ไม่มีชั้นวางลำดับนี้ในคลังสินค้า
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 03–04 | กำหนดค่า `sold_units`, `number_of_branches` | `sold_units = 40`, `number_of_branches = 0` | จำนวนสาขาเป็นศูนย์ |
| 06 | `average_sold = sold_units / number_of_branches` | - | คำนวณ `40 / 0` เกิด `ZeroDivisionError` |
| 09 | `print(...)` | - | ดักจับ `ZeroDivisionError` ได้สำเร็จ |
| 11 | `stock_input = "สามสิบ"` | `stock_input = "สามสิบ"` | ข้อความนี้ไม่ใช่ตัวเลข |
| 13 | `stock_count = int(stock_input)` | - | `int("สามสิบ")` แปลงไม่ได้ เกิด `ValueError` |
| 16 | `print(...)` | - | ดักจับ `ValueError` ได้สำเร็จ |
| 18–19 | กำหนดค่า `unit_price`, `discount_code` | `unit_price = 25` (int), `discount_code = "SALE10"` (str) | ชนิดข้อมูลต่างกัน |
| 21 | `discounted_price = unit_price + discount_code` | - | บวก `int` กับ `str` โดยตรง เกิด `TypeError` |
| 24 | `print(...)` | - | ดักจับ `TypeError` ได้สำเร็จ |
| 26 | `warehouse_shelves = ["A1", "A2", "A3"]` | สมาชิก 3 ตัว | ดัชนีที่ใช้ได้คือ 0, 1, 2 |
| 28 | `fourth_shelf = warehouse_shelves[3]` | - | เข้าถึงดัชนี 3 ซึ่งไม่มีอยู่จริง เกิด `IndexError` |
| 31 | `print(...)` | - | ดักจับ `IndexError` ได้สำเร็จ |

**คำอธิบายภาพรวม:**

ตัวอย่างนี้สาธิตข้อผิดพลาดทั้ง 4 ชนิดในบริบทของระบบสต็อกสินค้าร้านค้า โครงสร้างเหมือนกับตัวอย่างที่ 3.1 ทุก
ประการ เพียงเปลี่ยนสถานการณ์เป็นการคำนวณยอดขายเฉลี่ยต่อสาขา (บรรทัดที่ 3–9 นำไปสู่ `ZeroDivisionError`
เพราะจำนวนสาขาเป็นศูนย์) การแปลงจำนวนสต็อกจากข้อความภาษาไทย (บรรทัดที่ 11–16 นำไปสู่ `ValueError`)
การคำนวณราคาหลังหักส่วนลดโดยนำราคาตัวเลขไปบวกกับรหัสส่วนลดที่เป็นข้อความ (บรรทัดที่ 18–24 นำไปสู่
`TypeError`) และการเข้าถึงชั้นวางสินค้าในคลังที่ดัชนีเกินขอบเขต (บรรทัดที่ 26–31 นำไปสู่ `IndexError`) ทุกส่วน
ดักจับข้อผิดพลาดของตนเองได้สำเร็จ โปรแกรมจึงทำงานจนจบครบทั้ง 4 ส่วน

**จุดสังเกต:**

- สังเกตว่าโครงสร้างของโค้ดในตัวอย่างนี้เหมือนกับตัวอย่างที่ 3.1 ทุกประการ เปลี่ยนเพียงชื่อตัวแปรและข้อความ
  เท่านั้น ยืนยันว่าทั้ง 4 ชนิดข้อผิดพลาดนี้ไม่ได้ผูกติดกับโดเมนปัญหาใดโดยเฉพาะ แต่เกิดจากลักษณะการดำเนินการ
  กับข้อมูล (การหาร การแปลงชนิด การดำเนินการข้ามชนิด การเข้าถึงดัชนี) ซึ่งเกิดขึ้นได้ในทุกโดเมน
- ในบรรทัดที่ 21 หากสลับให้ `discount_code` เป็นตัวเลข (เช่น `10`) แทนที่จะเป็นข้อความ การบวกจะสำเร็จโดย
  ไม่มีข้อผิดพลาดใด ๆ เลย แสดงว่า `TypeError` ในกรณีนี้ขึ้นอยู่กับชนิดข้อมูลที่แท้จริง ไม่ใช่ค่าตัวเลขที่ปรากฏ
- การเข้าถึง `warehouse_shelves[3]` ในบรรทัดที่ 28 มีลักษณะเดียวกันกับตัวอย่างที่ 3.1 บรรทัดที่ 27 ทุกประการ
  คือพยายามเข้าถึงตำแหน่งที่ 4 ของรายการที่มีสมาชิกเพียง 3 ตัว ซึ่งเป็นรูปแบบความผิดพลาดที่พบบ่อยที่สุดของ
  `IndexError`

#### ตัวอย่างที่ 3.3 — ระบบจองที่นั่งเครื่องบิน (บริบทอื่นที่ต่างโดเมน)

In [9]:
print("สาธิตข้อผิดพลาดที่พบบ่อยในระบบจองที่นั่งเครื่องบิน")

total_passengers = 180
number_of_flights = 0
try:
    average_passengers = total_passengers / number_of_flights
    print("ผู้โดยสารเฉลี่ยต่อเที่ยวบินคือ", average_passengers)
except ZeroDivisionError:
    print("ZeroDivisionError: ไม่มีจำนวนเที่ยวบินให้คำนวณค่าเฉลี่ย")

seat_row_input = "แถวบี"
try:
    seat_row_number = int(seat_row_input)
    print("หมายเลขแถวที่นั่งคือ", seat_row_number)
except ValueError:
    print("ValueError: ไม่สามารถแปลงข้อความเป็นหมายเลขแถวที่นั่งได้")

ticket_price = 1200
passenger_count = "2"
try:
    total_fare = ticket_price + passenger_count
    print("ค่าโดยสารรวมคือ", total_fare)
except TypeError:
    print("TypeError: ไม่สามารถบวกราคาด้วยจำนวนผู้โดยสารที่เป็นข้อความได้")

boarding_gates = ["A1", "B2", "C3"]
try:
    fourth_gate = boarding_gates[3]
    print("ประตูขึ้นเครื่องที่สี่คือ", fourth_gate)
except IndexError:
    print("IndexError: ไม่มีประตูขึ้นเครื่องลำดับนี้")

สาธิตข้อผิดพลาดที่พบบ่อยในระบบจองที่นั่งเครื่องบิน
ZeroDivisionError: ไม่มีจำนวนเที่ยวบินให้คำนวณค่าเฉลี่ย
ValueError: ไม่สามารถแปลงข้อความเป็นหมายเลขแถวที่นั่งได้
TypeError: ไม่สามารถบวกราคาด้วยจำนวนผู้โดยสารที่เป็นข้อความได้
IndexError: ไม่มีประตูขึ้นเครื่องลำดับนี้


**Output (รันจริงยืนยันแล้ว):**
```
สาธิตข้อผิดพลาดที่พบบ่อยในระบบจองที่นั่งเครื่องบิน
ZeroDivisionError: ไม่มีจำนวนเที่ยวบินให้คำนวณค่าเฉลี่ย
ValueError: ไม่สามารถแปลงข้อความเป็นหมายเลขแถวที่นั่งได้
TypeError: ไม่สามารถบวกราคาด้วยจำนวนผู้โดยสารที่เป็นข้อความได้
IndexError: ไม่มีประตูขึ้นเครื่องลำดับนี้
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร-ค่า | คำอธิบายเพิ่มเติม |
|---|---|---|---|
| 03–04 | กำหนดค่า `total_passengers`, `number_of_flights` | `total_passengers = 180`, `number_of_flights = 0` | จำนวนเที่ยวบินเป็นศูนย์ |
| 06 | `average_passengers = total_passengers / number_of_flights` | - | คำนวณ `180 / 0` เกิด `ZeroDivisionError` |
| 09 | `print(...)` | - | ดักจับ `ZeroDivisionError` ได้สำเร็จ |
| 11 | `seat_row_input = "แถวบี"` | `seat_row_input = "แถวบี"` | ข้อความนี้ไม่ใช่ตัวเลข |
| 13 | `seat_row_number = int(seat_row_input)` | - | `int("แถวบี")` แปลงไม่ได้ เกิด `ValueError` |
| 16 | `print(...)` | - | ดักจับ `ValueError` ได้สำเร็จ |
| 18–19 | กำหนดค่า `ticket_price`, `passenger_count` | `ticket_price = 1200` (int), `passenger_count = "2"` (str) | ชนิดข้อมูลต่างกัน |
| 21 | `total_fare = ticket_price + passenger_count` | - | บวก `int` กับ `str` โดยตรง เกิด `TypeError` |
| 24 | `print(...)` | - | ดักจับ `TypeError` ได้สำเร็จ |
| 26 | `boarding_gates = ["A1", "B2", "C3"]` | สมาชิก 3 ตัว | ดัชนีที่ใช้ได้คือ 0, 1, 2 |
| 28 | `fourth_gate = boarding_gates[3]` | - | เข้าถึงดัชนี 3 ซึ่งไม่มีอยู่จริง เกิด `IndexError` |
| 31 | `print(...)` | - | ดักจับ `IndexError` ได้สำเร็จ |

**คำอธิบายภาพรวม:**

ตัวอย่างสุดท้ายของหัวข้อนี้สาธิตข้อผิดพลาดทั้ง 4 ชนิดในบริบทของระบบจองที่นั่งเครื่องบิน ซึ่งเป็นโดเมนที่แตกต่าง
จากตัวอย่างที่ 3.1 (ระบบตัดเกรด) และ 3.2 (ระบบสต็อกสินค้า) โดยสิ้นเชิง แต่ยังคงใช้รูปแบบเดียวกันคือแบ่งเป็น
4 ส่วนย่อยอิสระต่อกัน ส่วนแรก (บรรทัดที่ 3–9) คำนวณจำนวนผู้โดยสารเฉลี่ยต่อเที่ยวบินเมื่อไม่มีเที่ยวบินเลย ทำให้
เกิด `ZeroDivisionError` ส่วนที่สอง (บรรทัดที่ 11–16) พยายามแปลงข้อความ "แถวบี" เป็นหมายเลขแถวที่นั่ง ทำให้
เกิด `ValueError` ส่วนที่สาม (บรรทัดที่ 18–24) พยายามบวกราคาบัตรโดยสาร (ตัวเลข) กับจำนวนผู้โดยสาร (ข้อความ)
โดยตรง ทำให้เกิด `TypeError` และส่วนสุดท้าย (บรรทัดที่ 26–31) พยายามเข้าถึงประตูขึ้นเครื่องที่ดัชนีเกินขอบเขต
ทำให้เกิด `IndexError` ทุกส่วนดักจับข้อผิดพลาดของตนเองได้สำเร็จ โปรแกรมจึงทำงานจนจบครบทั้ง 4 ส่วนเช่นเดียว
กับตัวอย่างก่อนหน้า

**จุดสังเกต:**

- ตัวอย่างที่ 3.1, 3.2 และ 3.3 ใช้โครงสร้างโค้ดที่เหมือนกันทุกประการ แตกต่างเพียงชื่อตัวแปรและโดเมนของปัญหา
  การเห็นรูปแบบเดิมซ้ำในบริบทที่แตกต่างกันถึง 3 ครั้งช่วยตอกย้ำว่าทั้ง 4 ชนิดข้อผิดพลาดนี้เป็นแนวคิดทั่วไปที่ไม่ได้
  ผูกกับสถานการณ์ใดสถานการณ์หนึ่งเป็นการเฉพาะ
- สังเกตลำดับของทั้ง 4 ส่วนในทุกตัวอย่างของหัวข้อนี้ เรียงตามลำดับเดียวกันเสมอคือ `ZeroDivisionError`,
  `ValueError`, `TypeError`, แล้วจึงเป็น `IndexError` ซึ่งเป็นลำดับเดียวกับที่ปรากฏในตารางสรุปของหัวข้อนี้
- ค่า `passenger_count = "2"` ในบรรทัดที่ 19 เป็นตัวเลขที่ถูกเก็บไว้ในรูปข้อความ (str) มองผิวเผินอาจดูเหมือน
  ตัวเลข แต่ไพทอนถือว่าเป็นคนละชนิดข้อมูลกับ `int` อย่างเคร่งครัด การบวกโดยตรงจึงยังคงทำให้เกิด `TypeError`
  แม้เนื้อหาของข้อความจะเป็นตัวเลขก็ตาม

### ขอบเขตเนื้อหา

หัวข้อนี้ตรงกับเนื้อหาที่สอนใน Slide 4 ของ `slides/07_exception.md`

### เชื่อมโยงกับหัวข้ออื่น

ชนิดข้อผิดพลาดทั้ง 4 ที่อธิบายในหัวข้อนี้เป็นชนิดเดียวกันกับที่ปรากฏในตัวอย่างของหัวข้อที่ 2 (ซึ่งดักจับทีละชนิด)
และจะถูกนำไปใช้ร่วมกันในฟังก์ชันเดียวในหัวข้อที่ 4 ถัดไป

---

## หัวข้อที่ 4: จับหลาย exception ในฟังก์ชันเดียว (multiple except blocks)

### แนวคิด

เมื่อฟังก์ชันหนึ่งมีการดำเนินการหลายขั้นตอนที่แต่ละขั้นตอนอาจก่อให้เกิดข้อผิดพลาดคนละชนิดกัน แนวคิดทั่วไปคือ
การเตรียมทางรับมือไว้สำหรับข้อผิดพลาดแต่ละชนิดแยกจากกัน เพื่อให้ฟังก์ชันสามารถตอบสนองต่อสาเหตุของปัญหา
ได้อย่างเจาะจง แทนที่จะตอบสนองด้วยข้อความเดียวกันไม่ว่าจะเกิดปัญหาประเภทใดก็ตาม การแยกทางรับมือเช่นนี้
ช่วยให้ผู้ใช้งานฟังก์ชัน (หรือผู้พัฒนาที่กำลังตรวจสอบข้อผิดพลาด) เข้าใจได้ทันทีว่าปัญหาที่เกิดขึ้นมีสาเหตุมาจากอะไร

ในภาษาไพทอน ทำได้โดยการเขียนบล็อก `except` มากกว่าหนึ่งบล็อกต่อจากบล็อก `try` เดียวกัน โดยแต่ละบล็อก
`except` ระบุชื่อชนิดข้อผิดพลาดของตนเอง เมื่อเกิดข้อผิดพลาดขึ้นภายในบล็อก `try` ไพทอนจะตรวจสอบบล็อก
`except` ที่เขียนไว้ตามลำดับจากบนลงล่าง แล้วเลือกใช้บล็อกแรกที่มีชนิดตรงกับข้อผิดพลาดที่เกิดขึ้นจริง หากไม่มี
บล็อกใดตรงกันเลย ข้อผิดพลาดนั้นจะไม่ถูกดักจับ และจะแพร่กระจาย (propagate) ออกไปนอกฟังก์ชันเหมือนกับกรณี
ที่ไม่มีการดักจับใด ๆ เลย

### ตัวอย่าง Python

#### ตัวอย่างที่ 4.1 — ฟังก์ชันดึงคะแนนสอบจากรายการข้อความ

In [10]:
def get_exam_score(score_texts, index):
    try:
        scores = [int(text) for text in score_texts]
        total = sum(scores)
        average = total / len(scores)
        print("ค่าเฉลี่ยของชุดคะแนนนี้คือ", average)
        return scores[index]
    except ZeroDivisionError:
        return "ไม่มีคะแนนสำหรับคำนวณค่าเฉลี่ย"
    except ValueError:
        return "พบข้อความที่ไม่ใช่ตัวเลขในรายการคะแนน"
    except IndexError:
        return "ไม่มีคะแนนลำดับนี้ในรายการ"


print(get_exam_score(["80", "90", "70"], 1))
print(get_exam_score([], 0))
print(get_exam_score(["80", "เอ", "70"], 1))
print(get_exam_score(["80", "90", "70"], 5))

ค่าเฉลี่ยของชุดคะแนนนี้คือ 80.0
90
ไม่มีคะแนนสำหรับคำนวณค่าเฉลี่ย
พบข้อความที่ไม่ใช่ตัวเลขในรายการคะแนน
ค่าเฉลี่ยของชุดคะแนนนี้คือ 80.0
ไม่มีคะแนนลำดับนี้ในรายการ


**Output (รันจริงยืนยันแล้ว):**
```
ค่าเฉลี่ยของชุดคะแนนนี้คือ 80.0
90
ไม่มีคะแนนสำหรับคำนวณค่าเฉลี่ย
พบข้อความที่ไม่ใช่ตัวเลขในรายการคะแนน
ค่าเฉลี่ยของชุดคะแนนนี้คือ 80.0
ไม่มีคะแนนลำดับนี้ในรายการ
```

**ตารางการเปลี่ยนแปลงค่า:**

| การเรียกใช้ | อาร์กิวเมนต์ | บล็อกที่ทำงาน | ค่าที่ return | คำอธิบายเพิ่มเติม |
|---|---|---|---|---|
| ครั้งที่ 1 (บรรทัด 16) | `(["80", "90", "70"], 1)` | บล็อก `try` ทั้งหมดสำเร็จ | `90` | แปลงข้อความเป็นตัวเลขสำเร็จ คำนวณค่าเฉลี่ยได้ (80.0) และเข้าถึงดัชนี 1 ได้สำเร็จ |
| ครั้งที่ 2 (บรรทัด 17) | `([], 0)` | `except ZeroDivisionError` | ข้อความแจ้งเตือน | รายการว่างเปล่า `len(scores)` เป็น 0 คำนวณ `average` เกิดการหารด้วยศูนย์ในบรรทัดที่ 5 |
| ครั้งที่ 3 (บรรทัด 18) | `(["80", "เอ", "70"], 1)` | `except ValueError` | ข้อความแจ้งเตือน | `int("เอ")` ในบรรทัดที่ 3 แปลงไม่ได้ |
| ครั้งที่ 4 (บรรทัด 19) | `(["80", "90", "70"], 5)` | `except IndexError` | ข้อความแจ้งเตือน | คำนวณค่าเฉลี่ยสำเร็จ (พิมพ์ออกมาแล้ว) แต่ `scores[5]` ในบรรทัดที่ 7 เกินขอบเขต (มีสมาชิกเพียง 3 ตัว) |

**คำอธิบายภาพรวม:**

ฟังก์ชัน `get_exam_score` ในบรรทัดที่ 1–13 มีบล็อก `try` เดียว (บรรทัดที่ 2–7) ที่ทำงาน 3 ขั้นตอนต่อเนื่องกัน คือ
แปลงรายการข้อความให้เป็นตัวเลข (บรรทัดที่ 3) คำนวณค่าเฉลี่ยแล้วพิมพ์ออกมา (บรรทัดที่ 4–6) และคืนค่าคะแนน
ที่ตำแหน่งดัชนีที่ระบุ (บรรทัดที่ 7) โดยมีบล็อก `except` ถึง 3 บล็อก (บรรทัดที่ 8–13) แต่ละบล็อกดักจับข้อผิดพลาด
คนละชนิดที่อาจเกิดขึ้นได้จากขั้นตอนต่าง ๆ ในบล็อก `try` เมื่อเรียกใช้ฟังก์ชันนี้ 4 ครั้งด้วยอาร์กิวเมนต์ที่แตกต่างกัน
ในบรรทัดที่ 16–19 จะเห็นว่าการเรียกครั้งที่ 1 ทำงานสำเร็จทั้งหมด ในขณะที่การเรียกครั้งที่ 2, 3 และ 4 แต่ละครั้ง
กระตุ้นให้เกิดข้อผิดพลาดคนละชนิด และถูกดักจับโดยบล็อก `except` ที่สอดคล้องกันตามลำดับ ฟังก์ชันจึงคืนค่าออกมา
เป็นค่าเดียวเสมอไม่ว่าจะเกิดอะไรขึ้นภายใน ไม่เคยปล่อยให้โปรแกรมหยุดทำงานกลางคันเลยสักครั้ง

**จุดสังเกต:**

- แม้การเรียกครั้งที่ 4 (บรรทัดที่ 19) จะเกิด `IndexError` ในบรรทัดที่ 7 แต่ข้อความ "ค่าเฉลี่ยของชุดคะแนนนี้คือ
  80.0" ในบรรทัดที่ 6 ยังคงถูกพิมพ์ออกมาก่อน เพราะบรรทัดที่ 6 ทำงานสำเร็จไปแล้วก่อนที่จะถึงบรรทัดที่ 7 ซึ่งเป็น
  จุดเกิดปัญหาจริง แสดงว่าคำสั่งทุกบรรทัดก่อนจุดเกิดข้อผิดพลาดภายในบล็อก `try` เดียวกันยังคงทำงานตามปกติ
- ลำดับของบล็อก `except` ในบรรทัดที่ 8, 10 และ 12 ไม่มีผลต่อผลลัพธ์ในตัวอย่างนี้ เนื่องจากแต่ละชนิดข้อผิดพลาด
  ไม่ทับซ้อนกัน (แต่ละชนิดจะถูกเลือกโดยชนิดที่ตรงกันเท่านั้น) อย่างไรก็ตาม หากมีชนิดข้อผิดพลาดที่เป็นชนิดแม่
  ของกันและกัน ลำดับการเขียนจะมีผลต่อว่าบล็อกใดถูกเลือกก่อน (เป็นเนื้อหาระดับที่สูงกว่าขอบเขตของสัปดาห์นี้)
- ฟังก์ชันนี้ return ค่าเพียงค่าเดียวเสมอในทุกเส้นทางการทำงาน (ทั้งกรณีสำเร็จและกรณีเกิดข้อผิดพลาดแต่ละชนิด)
  สอดคล้องกับรูปแบบฟังก์ชันที่สืบทอดมาจากสัปดาห์ที่ 6 ซึ่งกำหนดว่าฟังก์ชันควร return ค่าเดียวเสมอ

#### ตัวอย่างที่ 4.2 — ฟังก์ชันคำนวณปริมาณวัตถุดิบต่อหนึ่งหน่วยบริโภค (บริบทอื่นที่ต่างโดเมน)

In [11]:
def get_scaled_ingredient(amount_texts, servings, index):
    try:
        amounts = [float(text) for text in amount_texts]
        total_amount = sum(amounts)
        amount_per_serving = total_amount / servings
        print("ปริมาณวัตถุดิบรวมต่อหนึ่งหน่วยบริโภคคือ", amount_per_serving)
        return amounts[index]
    except ZeroDivisionError:
        return "ไม่สามารถคำนวณได้เนื่องจากจำนวนหน่วยบริโภคเป็นศูนย์"
    except ValueError:
        return "พบข้อความที่ไม่ใช่ตัวเลขในรายการปริมาณวัตถุดิบ"
    except IndexError:
        return "ไม่มีวัตถุดิบลำดับนี้ในสูตรอาหาร"


print(get_scaled_ingredient(["200", "100", "50"], 4, 0))
print(get_scaled_ingredient(["200", "100", "50"], 0, 0))
print(get_scaled_ingredient(["200", "หนึ่งถ้วย", "50"], 4, 0))
print(get_scaled_ingredient(["200", "100", "50"], 4, 10))

ปริมาณวัตถุดิบรวมต่อหนึ่งหน่วยบริโภคคือ 87.5
200.0
ไม่สามารถคำนวณได้เนื่องจากจำนวนหน่วยบริโภคเป็นศูนย์
พบข้อความที่ไม่ใช่ตัวเลขในรายการปริมาณวัตถุดิบ
ปริมาณวัตถุดิบรวมต่อหนึ่งหน่วยบริโภคคือ 87.5
ไม่มีวัตถุดิบลำดับนี้ในสูตรอาหาร


**Output (รันจริงยืนยันแล้ว):**
```
ปริมาณวัตถุดิบรวมต่อหนึ่งหน่วยบริโภคคือ 87.5
200.0
ไม่สามารถคำนวณได้เนื่องจากจำนวนหน่วยบริโภคเป็นศูนย์
พบข้อความที่ไม่ใช่ตัวเลขในรายการปริมาณวัตถุดิบ
ปริมาณวัตถุดิบรวมต่อหนึ่งหน่วยบริโภคคือ 87.5
ไม่มีวัตถุดิบลำดับนี้ในสูตรอาหาร
```

**ตารางการเปลี่ยนแปลงค่า:**

| การเรียกใช้ | อาร์กิวเมนต์ | บล็อกที่ทำงาน | ค่าที่ return | คำอธิบายเพิ่มเติม |
|---|---|---|---|---|
| ครั้งที่ 1 (บรรทัด 16) | `(["200","100","50"], 4, 0)` | บล็อก `try` ทั้งหมดสำเร็จ | `200.0` | รวมปริมาณได้ 350.0 หารด้วย 4 หน่วยบริโภคได้ 87.5 แล้วคืนค่าที่ดัชนี 0 |
| ครั้งที่ 2 (บรรทัด 17) | `(["200","100","50"], 0, 0)` | `except ZeroDivisionError` | ข้อความแจ้งเตือน | `servings` เป็น 0 คำนวณในบรรทัดที่ 5 เกิดการหารด้วยศูนย์ |
| ครั้งที่ 3 (บรรทัด 18) | `(["200","หนึ่งถ้วย","50"], 4, 0)` | `except ValueError` | ข้อความแจ้งเตือน | `float("หนึ่งถ้วย")` ในบรรทัดที่ 3 แปลงไม่ได้ |
| ครั้งที่ 4 (บรรทัด 19) | `(["200","100","50"], 4, 10)` | `except IndexError` | ข้อความแจ้งเตือน | คำนวณสำเร็จ (พิมพ์ออกมาแล้ว) แต่ `amounts[10]` เกินขอบเขต (มีสมาชิกเพียง 3 ตัว) |

**คำอธิบายภาพรวม:**

ฟังก์ชัน `get_scaled_ingredient` มีโครงสร้างเดียวกันกับตัวอย่างที่ 4.1 ทุกประการ แต่เปลี่ยนบริบทเป็นการคำนวณ
ปริมาณวัตถุดิบต่อหนึ่งหน่วยบริโภคในสูตรอาหาร บล็อก `try` (บรรทัดที่ 2–7) แปลงรายการข้อความปริมาณวัตถุดิบให้
เป็นเลขทศนิยม รวมค่าทั้งหมด หารด้วยจำนวนหน่วยบริโภคที่ต้องการ แล้วคืนค่าปริมาณของวัตถุดิบชิ้นที่ระบุ การเรียก
ใช้ฟังก์ชันทั้ง 4 ครั้งในบรรทัดที่ 16–19 แสดงให้เห็นเส้นทางการทำงานที่แตกต่างกันทั้งหมด ทั้งกรณีสำเร็จและกรณี
เกิดข้อผิดพลาดแต่ละชนิด ซึ่งแต่ละกรณีถูกดักจับโดยบล็อก `except` ที่ตรงกับสาเหตุอย่างถูกต้อง

**จุดสังเกต:**

- โครงสร้างของฟังก์ชันนี้เหมือนกับตัวอย่างที่ 4.1 ทุกประการ ทั้งจำนวนบล็อก `except` ชนิดของข้อผิดพลาดที่ดักจับ
  และลำดับการเขียน แสดงว่ารูปแบบ "หนึ่งฟังก์ชัน หลาย except" นี้เป็นแม่แบบที่นำไปใช้ซ้ำได้กับปัญหาต่างโดเมน
  โดยเปลี่ยนเพียงรายละเอียดของขั้นตอนภายในบล็อก `try` เท่านั้น
- การเรียกครั้งที่ 4 (บรรทัดที่ 19) ยังคงพิมพ์ข้อความ "ปริมาณวัตถุดิบรวมต่อหนึ่งหน่วยบริโภคคือ 87.5" ออกมาก่อน
  จะเกิด `IndexError` เช่นเดียวกับที่พบในตัวอย่างที่ 4.1 ยืนยันหลักการเดิมว่าคำสั่งก่อนจุดเกิดปัญหาในบล็อก `try`
  ยังคงทำงานได้ตามปกติเสมอ
- หาก `servings` เป็นจำนวนลบแทนที่จะเป็นศูนย์ (เช่น -4) การหารจะไม่เกิด `ZeroDivisionError` แต่จะได้ผลลัพธ์
  เป็นค่าติดลบแทน ซึ่งเป็นค่าที่ผิดปกติในเชิงตรรกะแต่ไม่ใช่ข้อผิดพลาดที่ไพทอนตรวจพบให้อัตโนมัติ

#### ตัวอย่างที่ 4.3 — ฟังก์ชันคำนวณค่าคอมมิชชันยอดขายรายเดือน (บริบทอื่นที่ต่างโดเมน)

In [12]:
def get_monthly_commission(sales_texts, number_of_working_days, month_index):
    try:
        sales_values = [float(text) for text in sales_texts]
        total_sales = sum(sales_values)
        daily_average = total_sales / number_of_working_days
        print("ยอดขายเฉลี่ยต่อวันทำงานคือ", daily_average)
        return sales_values[month_index]
    except ZeroDivisionError:
        return "ไม่สามารถคำนวณได้เนื่องจากจำนวนวันทำงานเป็นศูนย์"
    except ValueError:
        return "พบข้อความที่ไม่ใช่ตัวเลขในรายการยอดขาย"
    except IndexError:
        return "ไม่มีข้อมูลยอดขายของเดือนลำดับนี้"


print(get_monthly_commission(["50000", "62000", "48000"], 26, 1))
print(get_monthly_commission(["50000", "62000", "48000"], 0, 1))
print(get_monthly_commission(["50000", "ไม่ระบุ", "48000"], 26, 1))
print(get_monthly_commission(["50000", "62000", "48000"], 26, 12))

ยอดขายเฉลี่ยต่อวันทำงานคือ 6153.846153846154
62000.0
ไม่สามารถคำนวณได้เนื่องจากจำนวนวันทำงานเป็นศูนย์
พบข้อความที่ไม่ใช่ตัวเลขในรายการยอดขาย
ยอดขายเฉลี่ยต่อวันทำงานคือ 6153.846153846154
ไม่มีข้อมูลยอดขายของเดือนลำดับนี้


**Output (รันจริงยืนยันแล้ว):**
```
ยอดขายเฉลี่ยต่อวันทำงานคือ 6153.846153846154
62000.0
ไม่สามารถคำนวณได้เนื่องจากจำนวนวันทำงานเป็นศูนย์
พบข้อความที่ไม่ใช่ตัวเลขในรายการยอดขาย
ยอดขายเฉลี่ยต่อวันทำงานคือ 6153.846153846154
ไม่มีข้อมูลยอดขายของเดือนลำดับนี้
```

**ตารางการเปลี่ยนแปลงค่า:**

| การเรียกใช้ | อาร์กิวเมนต์ | บล็อกที่ทำงาน | ค่าที่ return | คำอธิบายเพิ่มเติม |
|---|---|---|---|---|
| ครั้งที่ 1 (บรรทัด 16) | `(["50000","62000","48000"], 26, 1)` | บล็อก `try` ทั้งหมดสำเร็จ | `62000.0` | รวมยอดขายได้ 160000.0 หารด้วย 26 วันได้ประมาณ 6153.85 แล้วคืนค่าที่ดัชนี 1 |
| ครั้งที่ 2 (บรรทัด 17) | `(["50000","62000","48000"], 0, 1)` | `except ZeroDivisionError` | ข้อความแจ้งเตือน | `number_of_working_days` เป็น 0 คำนวณในบรรทัดที่ 5 เกิดการหารด้วยศูนย์ |
| ครั้งที่ 3 (บรรทัด 18) | `(["50000","ไม่ระบุ","48000"], 26, 1)` | `except ValueError` | ข้อความแจ้งเตือน | `float("ไม่ระบุ")` ในบรรทัดที่ 3 แปลงไม่ได้ |
| ครั้งที่ 4 (บรรทัด 19) | `(["50000","62000","48000"], 26, 12)` | `except IndexError` | ข้อความแจ้งเตือน | คำนวณสำเร็จ (พิมพ์ออกมาแล้ว) แต่ `sales_values[12]` เกินขอบเขต (มีสมาชิกเพียง 3 ตัว) |

**คำอธิบายภาพรวม:**

ฟังก์ชัน `get_monthly_commission` เป็นตัวอย่างที่สามในหัวข้อนี้ที่ใช้โครงสร้าง "หนึ่งฟังก์ชัน หลาย except" แบบ
เดียวกัน แต่เปลี่ยนบริบทเป็นระบบคำนวณค่าคอมมิชชันจากยอดขายรายเดือนของพนักงาน บล็อก `try` (บรรทัดที่ 2–7)
แปลงรายการข้อความยอดขายให้เป็นเลขทศนิยม รวมยอดขายทั้งหมด หารด้วยจำนวนวันทำงาน แล้วคืนค่ายอดขายของ
เดือนที่ระบุ การเรียกใช้ฟังก์ชันทั้ง 4 ครั้งในบรรทัดที่ 16–19 ยืนยันรูปแบบเดียวกับสองตัวอย่างก่อนหน้า คือมีเส้นทาง
สำเร็จ 1 เส้นทางและเส้นทางเกิดข้อผิดพลาดอีก 3 เส้นทางที่สอดคล้องกับบล็อก `except` แต่ละชนิด

**จุดสังเกต:**

- ตัวอย่างที่ 4.1, 4.2 และ 4.3 ยืนยันรูปแบบเดียวกันซ้ำถึง 3 ครั้งใน 3 โดเมนที่แตกต่างกันโดยสิ้นเชิง (การศึกษา
  การทำอาหาร และธุรกิจ) แสดงว่าโครงสร้าง "หนึ่งฟังก์ชัน หลาย except" เป็นรูปแบบทั่วไปที่ใช้ได้กับทุกสถานการณ์
  ที่ฟังก์ชันมีหลายขั้นตอนซึ่งแต่ละขั้นตอนอาจก่อให้เกิดข้อผิดพลาดคนละชนิด
- สังเกตว่าค่า return ในทุกเส้นทางของทั้ง 3 ตัวอย่างเป็นค่าเดียวเสมอ (ไม่มีการ return หลายค่าพร้อมกัน) แม้ชนิด
  ของค่าที่ return จะต่างกันไปในแต่ละเส้นทาง (ตัวเลขในกรณีสำเร็จ ข้อความในกรณีเกิดข้อผิดพลาด) ก็ตาม
- หากจำนวนวันทำงาน (`number_of_working_days`) ในบรรทัดที่ 17 เปลี่ยนจาก 0 เป็นจำนวนติดลบ ผลลัพธ์ของ
  การหารจะเป็นค่าติดลบแทนที่จะเกิด `ZeroDivisionError` ซึ่งเป็นสถานการณ์เดียวกับที่กล่าวถึงในจุดสังเกตของ
  ตัวอย่างที่ 4.2

### ขอบเขตเนื้อหา

หัวข้อนี้ต่อยอดจากฟังก์ชัน `safe_divide` ใน Slide 5 ของ `slides/07_exception.md` โดยขยายจากการดักจับ
ข้อผิดพลาดชนิดเดียวไปสู่การดักจับหลายชนิดในฟังก์ชันเดียวกัน

### เชื่อมโยงกับหัวข้ออื่น

รูปแบบฟังก์ชันในหัวข้อนี้ต่อยอดโดยตรงจากโครงสร้าง `try`/`except` พื้นฐานในหัวข้อที่ 2 และใช้ชนิดข้อผิดพลาดทั้ง 4
ที่อธิบายไว้ในหัวข้อที่ 3 หัวข้อถัดไป (หัวข้อที่ 5) จะนำแนวคิดฟังก์ชันที่มี `try`/`except` ภายในนี้ไปประยุกต์เป็น
"ฟังก์ชันเชิงป้องกัน" (defensive function) ที่เน้นการ return ค่าเดียวที่สอดคล้องกันเสมอ ตามรูปแบบของ
`safe_divide`

---

## หัวข้อที่ 5: Defensive Input/Function — เขียนฟังก์ชันให้ทนทานต่อ error

### แนวคิด

แนวคิดของการเขียนโปรแกรมเชิงป้องกัน (defensive programming) คือการออกแบบฟังก์ชันหรือส่วนของโปรแกรม
ให้สามารถรับมือกับข้อมูลนำเข้าที่ไม่ถูกต้องหรือสถานการณ์ที่ไม่คาดคิดได้โดยไม่ทำให้โปรแกรมทั้งระบบหยุดทำงาน
แนวคิดนี้ไม่ได้จำกัดเฉพาะการจัดการข้อผิดพลาดขณะรันเท่านั้น แต่ครอบคลุมถึงการตรวจสอบเงื่อนไขต่าง ๆ ก่อนที่จะ
ดำเนินการ อย่างไรก็ตาม ในบริบทของสัปดาห์นี้ การเขียนโปรแกรมเชิงป้องกันหมายถึงการนำ `try`/`except` มาห่อหุ้ม
การดำเนินการที่อาจก่อให้เกิดข้อผิดพลาดไว้ภายในฟังก์ชัน เพื่อให้ฟังก์ชันนั้น "ปลอดภัย" เสมอ กล่าวคือไม่ว่าจะเรียก
ใช้ด้วยข้อมูลนำเข้าแบบใดก็ตาม ฟังก์ชันจะคืนค่าออกมาเป็นค่าเดียวเสมอ ไม่ปล่อยให้ข้อผิดพลาดแพร่กระจายออกไป
ทำให้โปรแกรมที่เรียกใช้ฟังก์ชันนี้ต่อหยุดทำงานกลางคัน

รูปแบบนี้สอดคล้องโดยตรงกับฟังก์ชัน `safe_divide` ที่ปรากฏใน Slide 5 ของ `slides/07_exception.md` ซึ่งใช้
`try`/`except` ดักจับ `ZeroDivisionError` แล้วคืนค่า 0 แทนการปล่อยให้โปรแกรมหยุดทำงาน ฟังก์ชันในลักษณะนี้
มักถูกตั้งชื่อขึ้นต้นด้วยคำว่า `safe_` เพื่อสื่อความหมายว่าเป็นฟังก์ชันที่ "ปลอดภัย" ต่อการเรียกใช้ในทุกกรณี

### ตัวอย่าง Python

#### ตัวอย่างที่ 5.1 — ฟังก์ชันคำนวณค่าเฉลี่ยการเข้าเรียน (แนวคิดหลัก ต่อยอดจาก safe_divide)

In [13]:
def safe_get_average(scores):
    try:
        return sum(scores) / len(scores)
    except ZeroDivisionError:
        return 0


print(safe_get_average([80, 90, 70]))
print(safe_get_average([]))

80.0
0


**Output (รันจริงยืนยันแล้ว):**
```
80.0
0
```

**ตารางการเปลี่ยนแปลงค่า:**

| การเรียกใช้ | อาร์กิวเมนต์ | บล็อกที่ทำงาน | ค่าที่ return | คำอธิบายเพิ่มเติม |
|---|---|---|---|---|
| ครั้งที่ 1 (บรรทัด 8) | `[80, 90, 70]` | บล็อก `try` สำเร็จ | `80.0` | `sum([80,90,70]) = 240`, `len([80,90,70]) = 3`, `240/3 = 80.0` |
| ครั้งที่ 2 (บรรทัด 9) | `[]` | `except ZeroDivisionError` | `0` | `len([]) = 0` การหารในบรรทัดที่ 3 เกิด `ZeroDivisionError` จึงคืนค่า 0 แทน |

**คำอธิบายภาพรวม:**

ฟังก์ชัน `safe_get_average` ในบรรทัดที่ 1–5 มีโครงสร้างเดียวกันกับ `safe_divide` ใน Slide 5 ทุกประการ คือมี
บล็อก `try` เพียงบรรทัดเดียวที่ทำการคำนวณและ `return` ค่าออกไปทันทีหากสำเร็จ (บรรทัดที่ 3) และมีบล็อก
`except ZeroDivisionError` (บรรทัดที่ 4–5) ที่คืนค่า 0 แทนหากรายการที่รับเข้ามาว่างเปล่า เมื่อเรียกใช้ฟังก์ชันนี้
ด้วยรายการคะแนนที่มีสมาชิก (บรรทัดที่ 8) ฟังก์ชันจะคำนวณค่าเฉลี่ยได้ตามปกติและคืนค่า 80.0 ออกมา แต่เมื่อเรียก
ใช้ด้วยรายการว่างเปล่า (บรรทัดที่ 9) การคำนวณ `len([])` จะได้ 0 ทำให้เกิดการหารด้วยศูนย์ในบรรทัดที่ 3 แต่ด้วย
บล็อก `except` ที่เตรียมไว้ ฟังก์ชันจึงคืนค่า 0 ออกมาแทนที่จะปล่อยให้โปรแกรมหยุดทำงาน

**จุดสังเกต:**

- ฟังก์ชันนี้ return ค่าเพียงค่าเดียวเสมอในทุกกรณี (ทั้ง `80.0` และ `0`) ไม่มีเส้นทางใดที่ return มากกว่าหนึ่งค่า
  หรือไม่ return เลย ซึ่งเป็นคุณสมบัติสำคัญของฟังก์ชันเชิงป้องกันตามรูปแบบที่สืบทอดมาจาก `safe_divide`
- สังเกตว่าค่าที่ return ในกรณีสำเร็จ (`80.0`) เป็นชนิดทศนิยม (`float`) ในขณะที่ค่าที่ return ในกรณีเกิด
  ข้อผิดพลาด (`0`) เป็นชนิดจำนวนเต็ม (`int`) ความแตกต่างของชนิดข้อมูลระหว่างสองเส้นทางนี้เป็นลักษณะเดียว
  กับที่ปรากฏใน `safe_divide` ต้นแบบใน Slide 5 อยู่แล้ว
- ฟังก์ชันนี้ดักจับเฉพาะ `ZeroDivisionError` เพียงชนิดเดียวเท่านั้น ความปลอดภัยของฟังก์ชันจึงจำกัดอยู่เฉพาะ
  กรณีที่รายการว่างเปล่า หากเรียกใช้ด้วยข้อมูลนำเข้าที่ก่อให้เกิดข้อผิดพลาดชนิดอื่น (เช่น ส่งค่าที่ไม่ใช่รายการเข้ามา)
  ฟังก์ชันนี้จะยังคงไม่ปลอดภัยจากข้อผิดพลาดชนิดนั้น ดังจะเห็นตัวอย่างที่ชัดเจนยิ่งขึ้นในตัวอย่างที่ 5.3

#### ตัวอย่างที่ 5.2 — ฟังก์ชันแปลงค่าอุณหภูมิจากข้อความ (บริบทอื่นที่ต่างโดเมน)

In [14]:
def safe_convert_temperature(text):
    try:
        return float(text)
    except ValueError:
        return None


print(safe_convert_temperature("32.5"))
print(safe_convert_temperature("ไม่ทราบค่า"))

32.5
None


**Output (รันจริงยืนยันแล้ว):**
```
32.5
None
```

**ตารางการเปลี่ยนแปลงค่า:**

| การเรียกใช้ | อาร์กิวเมนต์ | บล็อกที่ทำงาน | ค่าที่ return | คำอธิบายเพิ่มเติม |
|---|---|---|---|---|
| ครั้งที่ 1 (บรรทัด 8) | `"32.5"` | บล็อก `try` สำเร็จ | `32.5` | `float("32.5")` แปลงสำเร็จ ได้ค่าทศนิยม 32.5 |
| ครั้งที่ 2 (บรรทัด 9) | `"ไม่ทราบค่า"` | `except ValueError` | `None` | `float("ไม่ทราบค่า")` แปลงไม่ได้ เกิด `ValueError` จึงคืนค่า `None` แทน |

**คำอธิบายภาพรวม:**

ฟังก์ชัน `safe_convert_temperature` เปลี่ยนโดเมนจากการคำนวณค่าเฉลี่ยในตัวอย่างที่ 5.1 มาเป็นการแปลงข้อความ
เป็นค่าอุณหภูมิแบบทศนิยม โครงสร้างยังคงเป็น `try`/`except` เพียงคู่เดียว (บรรทัดที่ 2–5) โดยบล็อก `try` พยายาม
แปลงข้อความที่รับเข้ามาด้วย `float(...)` ในบรรทัดที่ 3 หากแปลงสำเร็จก็ `return` ค่าทศนิยมนั้นออกไปทันที แต่หาก
ข้อความที่รับเข้ามาไม่สามารถแปลงเป็นตัวเลขได้ (เช่น "ไม่ทราบค่า" ในบรรทัดที่ 9) จะเกิด `ValueError` ซึ่งถูกดักจับ
โดยบล็อก `except` ในบรรทัดที่ 4–5 และคืนค่า `None` แทน แทนที่จะคืนค่า 0 เหมือนในตัวอย่างที่ 5.1

**จุดสังเกต:**

- ตัวอย่างนี้เลือกคืนค่า `None` แทนที่จะเป็นตัวเลข 0 เมื่อเกิดข้อผิดพลาด ต่างจากตัวอย่างที่ 5.1 ที่คืนค่า 0 ทั้งนี้
  เพราะค่าอุณหภูมิ 0 มีความหมายทางกายภาพจริง (ศูนย์องศา) การใช้ 0 แทนกรณีข้อมูลผิดพลาดอาจทำให้ผู้ใช้เข้าใจ
  ผิดว่าอุณหภูมิจริงเป็นศูนย์องศา จึงเลือกใช้ `None` ซึ่งเป็นค่าที่สื่อความหมายว่า "ไม่มีค่า" อย่างชัดเจนกว่า
- แม้ค่าที่ return จะเปลี่ยนจาก 0 เป็น `None` แต่ฟังก์ชันนี้ยังคง return ค่าเพียงค่าเดียวเสมอในทุกเส้นทาง ซึ่ง
  เป็นคุณสมบัติเดียวกับตัวอย่างที่ 5.1 และฟังก์ชัน `safe_divide` ต้นแบบ
- ฟังก์ชันนี้ดักจับเฉพาะ `ValueError` เท่านั้น เช่นเดียวกับตัวอย่างที่ 5.1 ที่ดักจับเฉพาะ `ZeroDivisionError`
  แสดงว่าฟังก์ชันเชิงป้องกันแต่ละฟังก์ชันมักถูกออกแบบให้ป้องกันข้อผิดพลาดเฉพาะชนิดที่คาดว่าจะเกิดขึ้นได้จาก
  การดำเนินการภายในของตนเองเท่านั้น

#### ตัวอย่างที่ 5.3 — ฟังก์ชันแปลงอายุจากข้อความ (edge case: ฟังก์ชันเชิงป้องกันที่ยังไม่ครอบคลุมทุกกรณี)

In [15]:
def safe_convert_age(text):
    try:
        return int(text)
    except ValueError:
        return None


print(safe_convert_age("25"))
print(safe_convert_age("ยี่สิบห้า"))
print(safe_convert_age(None))

25
None


TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'

**Output (รันจริงยืนยันแล้ว):**
```
25
None
Traceback (most recent call last):
  File "exercise.py", line 10, in <module>
    print(safe_convert_age(None))
          ^^^^^^^^^^^^^^^^^^^^^^
  File "exercise.py", line 3, in safe_convert_age
    return int(text)
           ^^^^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

**ตารางการเปลี่ยนแปลงค่า:**

| การเรียกใช้ | อาร์กิวเมนต์ | บล็อกที่ทำงาน | ค่าที่ return | คำอธิบายเพิ่มเติม |
|---|---|---|---|---|
| ครั้งที่ 1 (บรรทัด 8) | `"25"` | บล็อก `try` สำเร็จ | `25` | `int("25")` แปลงสำเร็จ ได้จำนวนเต็ม 25 |
| ครั้งที่ 2 (บรรทัด 9) | `"ยี่สิบห้า"` | `except ValueError` | `None` | `int("ยี่สิบห้า")` แปลงไม่ได้ เกิด `ValueError` จึงคืนค่า `None` แทน |
| ครั้งที่ 3 (บรรทัด 10) | `None` | ไม่มีบล็อกใดดักจับได้ | ไม่มีการ return ฟังก์ชันหยุดทำงานกลางคัน | `int(None)` ก่อให้เกิด `TypeError` ไม่ใช่ `ValueError` บล็อก `except ValueError` ในบรรทัดที่ 4 จึงไม่ตรงชนิด ข้อผิดพลาดจึงแพร่กระจายออกจากฟังก์ชันไปยังจุดที่เรียกใช้ ทำให้โปรแกรมหยุดทำงานทั้งหมด |

**คำอธิบายภาพรวม:**

ฟังก์ชัน `safe_convert_age` มีโครงสร้างเดียวกันกับตัวอย่างที่ 5.1 และ 5.2 ทุกประการ คือมี `try`/`except`
เพียงคู่เดียวที่ดักจับ `ValueError` เมื่อเรียกใช้ด้วยข้อความที่เป็นตัวเลข ("25" ในบรรทัดที่ 8) ฟังก์ชันทำงานสำเร็จ
และคืนค่าจำนวนเต็ม 25 เมื่อเรียกใช้ด้วยข้อความที่ไม่ใช่ตัวเลข ("ยี่สิบห้า" ในบรรทัดที่ 9) เกิด `ValueError` ซึ่งถูก
ดักจับได้ตามที่ออกแบบไว้ และคืนค่า `None` แทน แต่เมื่อเรียกใช้ฟังก์ชันเดียวกันนี้ด้วยค่า `None` แทนที่จะเป็นข้อความ
(บรรทัดที่ 10) ฟังก์ชัน `int(...)` ในบรรทัดที่ 3 จะก่อให้เกิดข้อผิดพลาดชนิด `TypeError` แทน เนื่องจากบล็อก
`except` ในบรรทัดที่ 4 ระบุไว้เฉพาะ `ValueError` เท่านั้น จึงไม่สามารถดักจับ `TypeError` นี้ได้ ข้อผิดพลาดจึง
แพร่กระจายออกจากฟังก์ชัน `safe_convert_age` ไปยังจุดที่เรียกใช้งานในบรรทัดที่ 10 โดยตรง และทำให้โปรแกรม
ทั้งหมดหยุดทำงานกลางคัน

**จุดสังเกต:**

- ชื่อฟังก์ชัน `safe_convert_age` สื่อความหมายว่าเป็นฟังก์ชันที่ "ปลอดภัย" แต่จากการเรียกใช้ครั้งที่ 3 แสดงให้เห็น
  ว่าฟังก์ชันนี้ปลอดภัยเฉพาะจากข้อผิดพลาดชนิด `ValueError` เท่านั้น ไม่ได้ปลอดภัยจากข้อผิดพลาดทุกชนิดที่อาจ
  เกิดขึ้นได้จากการดำเนินการภายในบล็อก `try`
- การมีบล็อก `try`/`except` อยู่ภายในฟังก์ชันไม่ได้รับประกันว่าฟังก์ชันนั้นจะไม่ทำให้โปรแกรมหยุดทำงานเลยใน
  ทุกกรณี ความปลอดภัยของฟังก์ชันขึ้นอยู่กับว่าบล็อก `except` ที่เขียนไว้ครอบคลุมชนิดข้อผิดพลาดที่อาจเกิดขึ้นจริง
  ได้มากน้อยเพียงใด
- เปรียบเทียบกับตัวอย่างที่ 5.1 ที่ดักจับเฉพาะ `ZeroDivisionError` และตัวอย่างที่ 5.2 ที่ดักจับเฉพาะ `ValueError`
  ทั้งสองตัวอย่างก็มีข้อจำกัดในลักษณะเดียวกันนี้เช่นกัน หากถูกเรียกใช้ด้วยข้อมูลนำเข้าที่ก่อให้เกิดข้อผิดพลาดชนิดอื่น
  ที่ไม่ได้ถูกระบุไว้ในบล็อก `except` ของตน ก็จะเกิดผลลัพธ์แบบเดียวกันกับที่ปรากฏในตัวอย่างนี้

### ขอบเขตเนื้อหา

หัวข้อนี้ตรงกับเนื้อหาที่สอนใน Slide 5 ของ `slides/07_exception.md` (ฟังก์ชัน `safe_divide`) โดยขยายความ
ด้วยตัวอย่างเพิ่มเติมอีก 2 โดเมน และเพิ่มตัวอย่างที่ 5.3 ซึ่งแสดงขอบเขตของแนวคิด "ฟังก์ชันเชิงป้องกัน" ว่าครอบคลุม
เฉพาะชนิดข้อผิดพลาดที่ถูกระบุไว้ในบล็อก `except` เท่านั้น

### เชื่อมโยงกับหัวข้ออื่น

หัวข้อนี้เป็นการนำโครงสร้างพื้นฐานจากหัวข้อที่ 2 (try/except จับชนิดเดียว) และหัวข้อที่ 4 (จับหลายชนิดในฟังก์ชัน
เดียว) มาประยุกต์ใช้ในรูปแบบฟังก์ชันที่เน้นการคืนค่าเดียวที่สอดคล้องกันเสมอ ทั้งสามตัวอย่างในหัวข้อนี้ (5.1, 5.2,
5.3) สามารถนำไปเปรียบเทียบกับฟังก์ชันในตัวอย่างที่ 4.1–4.3 ได้โดยตรง เนื่องจากฟังก์ชันในหัวข้อที่ 4 ดักจับ
ข้อผิดพลาดหลายชนิดพร้อมกัน ในขณะที่ฟังก์ชันในหัวข้อนี้เน้นการดักจับข้อผิดพลาดเพียงชนิดเดียวต่อหนึ่งฟังก์ชัน
ตามรูปแบบดั้งเดิมของ `safe_divide`

---

## แบบฝึกหัดท้ายบท

แบบฝึกหัดต่อไปนี้มีทั้งหมด 20 ข้อ แบ่งเป็น 5 หัวข้อ หัวข้อละ 4 ข้อ ผสมผสานโจทย์ 4 ประเภท ได้แก่ การอธิบาย
แนวคิด การทายผลลัพธ์ การหาข้อผิดพลาด และการเขียนโค้ด สถานการณ์ที่ใช้ในแบบฝึกหัดทุกข้อออกแบบให้แตกต่าง
จากตัวอย่างในเนื้อหาข้างต้น เพื่อทดสอบความเข้าใจในหลักการ ไม่ใช่การจดจำโค้ดเดิม

### หัวข้อที่ 1: Exception คืออะไร ต่างจาก Syntax Error อย่างไร (ข้อ 1–4)

**ข้อ 1 (อธิบายแนวคิด)** จงอธิบายความแตกต่างระหว่างข้อผิดพลาดทางไวยากรณ์ (syntax error) กับข้อผิดพลาด
ขณะรัน (exception) โดยระบุว่าข้อผิดพลาดแต่ละประเภทถูกตรวจพบในช่วงเวลาใดของการทำงานของโปรแกรม

**ข้อ 2 (ทายผลลัพธ์)** กำหนดโค้ดต่อไปนี้ จงทายผลลัพธ์ที่จะปรากฏบนหน้าจอ
```python
01  queue_numbers = [101, 102, 103]
02  print("คิวถัดไปคือ", queue_numbers[0])
03  print("คิวคือ", queue_numbers[3])
04  print("จบการทำงาน")
```

**ข้อ 3 (หาข้อผิดพลาด)** โค้ดต่อไปนี้มีปัญหาอยู่ จงระบุว่าปัญหาคืออะไร และอธิบายว่าเหตุใดปัญหานี้จึงถูกตรวจพบ
ก่อนที่โปรแกรมจะเริ่มทำงานแม้แต่บรรทัดเดียว
```python
01  book_count = 12
02    print("จำนวนหนังสือคือ", book_count)
```

**ข้อ 4 (เขียนโค้ด)** จงเขียนโปรแกรมสั้น ๆ ที่จงใจทำให้เกิด `TypeError` โดยไม่มีการดักจับข้อผิดพลาดใด ๆ
(unhandled) แล้วอธิบายว่า traceback ที่ได้บอกข้อมูลอะไรบ้าง

In [16]:
# เขียนโค้ดของคุณสำหรับข้อ 4 ที่นี่


### หัวข้อที่ 2: try/except พื้นฐาน (ข้อ 5–8)

**ข้อ 5 (อธิบายแนวคิด)** จงอธิบายว่าเหตุใดคำสั่ง `except` จึงต้องระบุชื่อชนิดข้อผิดพลาดต่อท้าย และจะเกิดอะไรขึ้น
หากชนิดที่ระบุไว้ไม่ตรงกับชนิดข้อผิดพลาดที่เกิดขึ้นจริง

**ข้อ 6 (ทายผลลัพธ์)** กำหนดโค้ดต่อไปนี้ จงทายผลลัพธ์ที่จะปรากฏบนหน้าจอ
```python
01  fee_text = "ยี่สิบบาท"
02  try:
03      fee = int(fee_text)
04      print("ค่าจอดรถคือ", fee, "บาท")
05  except ValueError:
06      print("ข้อมูลค่าจอดรถไม่ถูกต้อง")
07  print("ระบบจอดรถทำงานต่อ")
```

**ข้อ 7 (หาข้อผิดพลาด)** โค้ดต่อไปนี้ตั้งใจจะดักจับข้อผิดพลาด แต่กลับทำให้โปรแกรมหยุดทำงานอยู่ดี จงระบุว่าเหตุใด
บล็อก `except` จึงไม่สามารถดักจับข้อผิดพลาดที่เกิดขึ้นได้
```python
01  distance_km = 10
02  try:
03      fuel_needed = distance_km / "0"
04      print(fuel_needed)
05  except ZeroDivisionError:
06      print("ระยะทางเป็นศูนย์")
```

**ข้อ 8 (เขียนโค้ด)** จงเขียนฟังก์ชันชื่อ `average_water_usage(total_liters, days)` ที่ใช้ `try`/`except`
ดักจับ `ZeroDivisionError` โดยคืนค่าข้อความแจ้งเตือนแทนเมื่อ `days` เป็นศูนย์ แล้วทดสอบเรียกใช้ด้วยค่า
`(300, 3)` และ `(300, 0)`

In [17]:
# เขียนโค้ดของคุณสำหรับข้อ 8 ที่นี่


### หัวข้อที่ 3: Exception types ที่พบบ่อย (ข้อ 9–12)

**ข้อ 9 (อธิบายแนวคิด)** จงจับคู่ชนิดข้อผิดพลาดทั้ง 4 ชนิด (`ZeroDivisionError`, `ValueError`, `TypeError`,
`IndexError`) กับสถานการณ์ที่ทำให้เกิดข้อผิดพลาดชนิดนั้น พร้อมอธิบายความแตกต่างระหว่าง `ValueError` กับ
`TypeError` ที่มักถูกสับสนกัน

**ข้อ 10 (ทายผลลัพธ์)** กำหนดโค้ดต่อไปนี้ จงทายผลลัพธ์ที่จะปรากฏบนหน้าจอ
```python
01  ticket_prices = [500, 750, 1000]
02  try:
03      chosen_price = ticket_prices[5]
04      print("ราคาบัตรที่เลือกคือ", chosen_price)
05  except IndexError:
06      print("ไม่มีบัตรราคานี้")
07
08  age_text = "สิบแปด"
09  try:
10      age = int(age_text)
11      print("อายุคือ", age)
12  except ValueError:
13      print("ข้อมูลอายุไม่ถูกต้อง")
```

**ข้อ 11 (หาข้อผิดพลาด)** โค้ดต่อไปนี้ทำให้โปรแกรมหยุดทำงานทั้งที่มีการดักจับข้อผิดพลาดไว้แล้ว จงระบุว่าชนิด
ข้อผิดพลาดที่เกิดขึ้นจริงคือชนิดใด และเพราะเหตุใดบล็อก `except` ที่เขียนไว้จึงไม่สามารถดักจับได้
```python
01  room_price = 1200
02  guest_count = "two"
03  try:
04      total_price = room_price + guest_count
05      print(total_price)
06  except ValueError:
07      print("ข้อมูลจำนวนผู้เข้าพักไม่ถูกต้อง")
```

**ข้อ 12 (เขียนโค้ด)** จงเขียนโปรแกรมสั้น ๆ ที่ใช้รายการ (list) ชื่อ `attendance` เก็บสถานะการเข้าเรียน 3 รายการ
แล้วจงใจเข้าถึงตำแหน่งดัชนีที่ 10 โดยไม่มีการดักจับข้อผิดพลาด เพื่อให้เกิด `IndexError` ขึ้นจริง

In [18]:
# เขียนโค้ดของคุณสำหรับข้อ 12 ที่นี่


### หัวข้อที่ 4: จับหลาย exception ในฟังก์ชันเดียว (ข้อ 13–16)

**ข้อ 13 (อธิบายแนวคิด)** จงอธิบายว่าเหตุใดการเขียนบล็อก `except` หลายบล็อกต่อจากบล็อก `try` เดียวกันจึงมี
ประโยชน์มากกว่าการใช้บล็อก `except` เพียงบล็อกเดียวที่รับมือกับทุกข้อผิดพลาดด้วยข้อความเดียวกัน

**ข้อ 14 (ทายผลลัพธ์)** กำหนดฟังก์ชันต่อไปนี้ จงทายผลลัพธ์ของการเรียกใช้ทั้ง 2 ครั้ง
```python
01  def get_fuel_cost(price_texts, distance, index):
02      try:
03          prices = [float(text) for text in price_texts]
04          total = sum(prices)
05          cost_per_km = total / distance
06          print("ต้นทุนต่อกิโลเมตรคือ", cost_per_km)
07          return prices[index]
08      except ZeroDivisionError:
09          return "ระยะทางเป็นศูนย์"
10      except ValueError:
11          return "พบข้อความที่ไม่ใช่ตัวเลขในรายการราคาน้ำมัน"
12      except IndexError:
13          return "ไม่มีราคาลำดับนี้"
14
15
16  print(get_fuel_cost(["30", "32", "31"], 100, 0))
17  print(get_fuel_cost(["30", "32", "31"], 0, 0))
```

**ข้อ 15 (หาข้อผิดพลาด)** ฟังก์ชันต่อไปนี้ทำงานได้ตามปกติในการเรียกใช้ครั้งแรก แต่ทำให้โปรแกรมหยุดทำงานใน
การเรียกใช้ครั้งที่สอง จงระบุว่าชนิดข้อผิดพลาดใดเกิดขึ้น และเพราะเหตุใดฟังก์ชันนี้จึงไม่สามารถดักจับได้
```python
01  def get_temperature_report(temp_texts, day_index):
02      try:
03          temps = [float(t) for t in temp_texts]
04          average = sum(temps) / len(temps)
05          print("อุณหภูมิเฉลี่ยคือ", average)
06          return temps[day_index]
07      except ZeroDivisionError:
08          return "ไม่มีข้อมูลอุณหภูมิ"
09      except ValueError:
10          return "พบข้อความที่ไม่ใช่ตัวเลขในรายการอุณหภูมิ"
11
12
13  print(get_temperature_report(["30.5", "31.0", "29.8"], 1))
14  print(get_temperature_report(["30.5", "31.0", "29.8"], 10))
```

**ข้อ 16 (เขียนโค้ด)** จงเขียนฟังก์ชันชื่อ `get_medicine_stock(stock_texts, index)` สำหรับระบบสต็อกยาใน
ร้านขายยา ที่มีบล็อก `try` เดียวคู่กับบล็อก `except` อย่างน้อย 2 บล็อก เพื่อดักจับทั้ง `ValueError` (กรณีข้อความ
ในรายการแปลงเป็นตัวเลขไม่ได้) และ `IndexError` (กรณีดัชนีเกินขอบเขต) แล้วทดสอบเรียกใช้อย่างน้อย 3 กรณี

In [19]:
# เขียนโค้ดของคุณสำหรับข้อ 16 ที่นี่


### หัวข้อที่ 5: Defensive Input/Function (ข้อ 17–20)

**ข้อ 17 (อธิบายแนวคิด)** จงอธิบายว่าฟังก์ชันเชิงป้องกัน (defensive function) คืออะไร และมีความสัมพันธ์อย่างไร
กับฟังก์ชัน `safe_divide` ใน Slide 5 ของ `slides/07_exception.md`

**ข้อ 18 (ทายผลลัพธ์)** กำหนดฟังก์ชันต่อไปนี้ จงทายผลลัพธ์ของการเรียกใช้ทั้ง 2 ครั้ง
```python
01  def safe_get_discount_percent(original_price, discounted_price):
02      try:
03          return (original_price - discounted_price) / original_price * 100
04      except ZeroDivisionError:
05          return 0
06
07
08  print(safe_get_discount_percent(500, 400))
09  print(safe_get_discount_percent(0, 0))
```

**ข้อ 19 (หาข้อผิดพลาด)** ฟังก์ชันต่อไปนี้ถูกตั้งชื่อว่า "safe" แต่การเรียกใช้ครั้งที่สามกลับทำให้โปรแกรมหยุดทำงาน
จงระบุว่าเหตุใดจึงเป็นเช่นนั้น
```python
01  def safe_convert_weight(text):
02      try:
03          return float(text)
04      except ValueError:
05          return None
06
07
08  print(safe_convert_weight("65.5"))
09  print(safe_convert_weight("หกสิบห้า"))
10  print(safe_convert_weight(None))
```

**ข้อ 20 (เขียนโค้ด)** จงเขียนฟังก์ชันชื่อ `safe_get_item_price(price_list, index)` สำหรับระบบราคาสินค้า ที่
ใช้ `try`/`except` ดักจับ `IndexError` โดยคืนค่า `None` แทนเมื่อดัชนีเกินขอบเขต แล้วทดสอบเรียกใช้ด้วยรายการ
`[100, 200, 300]` ที่ดัชนี 1 และดัชนี 10

---

In [20]:
# เขียนโค้ดของคุณสำหรับข้อ 20 ที่นี่


> **ข้อมูลเพิ่มเติม:** เฉลยของแบบฝึกหัดทั้ง 20 ข้อในสมุดบันทึกนี้ถูกย้ายไปไว้ในไฟล์ `lecture_note_basic_answers.ipynb` (อยู่ในโฟลเดอร์เดียวกัน) เปิดไฟล์นั้นเมื่อทำแบบฝึกหัดเสร็จแล้วและต้องการตรวจคำตอบ